From 57c5affaa8618a064c123e42db8e69d186e80c4e Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 08:40:36 +0800
Subject: [PATCH] Fix Torch version


diff --git a/pom.xml b/pom.xml
index 4dccd64..b34583e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -11,7 +11,7 @@
 
     <groupId>net.techcable.tacospigot</groupId>
     <artifactId>tacospigot-api</artifactId>
-    <version>1.9.4-R0.1</version>
+    <version>1.9.4-R1.2</version>
     <packaging>jar</packaging>
 
     <name>TacoSpigot-API</name>
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 32177d3..73b144f 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -192,8 +192,8 @@ public class VersionCommand extends BukkitCommand {
         String version = Bukkit.getVersion();
         if (version == null) version = "Custom";
         // TacoSpigot start
-        if (version.startsWith("git-TacoSpigot-")) {
-            String[] parts = version.substring("git-TacoSpigot-".length()).split("[-\\s]");
+        if (version.startsWith("git-Torch-")) {
+            String[] parts = version.substring("git-TorchSpigot-".length()).split("[-\\s]");
             // TacoSpigot end
             int distance = getDistance(null, parts[0]);
             switch (distance) {
@@ -253,7 +253,7 @@ public class VersionCommand extends BukkitCommand {
         } catch (NumberFormatException ex) {
             */
             verInfo = verInfo.replace("\"", "");
-            return getFromRepo("TacoSpigot/TacoSpigot", verInfo); // 'PaperMC/Paper' -> 'TacoSpigot/TacoSpigot'
+            return getFromRepo("TorchSpigot/Torch", verInfo); // 'PaperMC/Paper' -> 'TacoSpigot/TacoSpigot'
         //}
         // TacoSpigot end
             /*
-- 
2.8.2.windows.1

