From 6075bd0414ef0ff09576db315b63e25de7f69285 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 14:51:52 +0800
Subject: [PATCH] MC-80966 - Always send chunk sections


diff --git a/src/main/java/net/minecraft/server/ChunkSection.java b/src/main/java/net/minecraft/server/ChunkSection.java
index 6cf71b8..94aea80 100644
--- a/src/main/java/net/minecraft/server/ChunkSection.java
+++ b/src/main/java/net/minecraft/server/ChunkSection.java
@@ -64,7 +64,9 @@ public class ChunkSection {
     }
 
     public boolean a() {
-        return this.nonEmptyBlockCount == 0;
+        // Paper - MC-80966
+        // Even if there are no blocks, there may be other information associated with the chunk, always send it.
+        return false;
     }
 
     public boolean shouldTick() {
-- 
2.8.2.windows.1

