From 53bae171b43cdd5ff8bcd5d12a78cdb2afd94514 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 00:50:24 +0800
Subject: [PATCH] torch signed


diff --git a/src/main/java/co/aikar/timings/MinecraftTimings.java b/src/main/java/co/aikar/timings/MinecraftTimings.java
index 0f8315c..744c5d4 100644
--- a/src/main/java/co/aikar/timings/MinecraftTimings.java
+++ b/src/main/java/co/aikar/timings/MinecraftTimings.java
@@ -1,5 +1,5 @@
 package co.aikar.timings;
-
+//Torch
 import net.minecraft.server.*;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.scheduler.BukkitTask;
-- 
2.8.2.windows.1

