From 9a589b3bacdd1d79a67771bac1f596d90973570e Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 15:20:37 +0800
Subject: [PATCH] SPIGOT-1132: Duplicate Minecart interact event


diff --git a/src/main/java/net/minecraft/server/ItemMinecart.java b/src/main/java/net/minecraft/server/ItemMinecart.java
index 0563f34..4b75cc8 100644
--- a/src/main/java/net/minecraft/server/ItemMinecart.java
+++ b/src/main/java/net/minecraft/server/ItemMinecart.java
@@ -109,13 +109,7 @@ public class ItemMinecart extends Item {
                     d0 = 0.5D;
                 }
 
-                // CraftBukkit start - Minecarts
-                org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(entityhuman, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, blockposition, enumdirection, itemstack, enumhand);
-
-                if (event.isCancelled()) {
-                    return EnumInteractionResult.PASS;
-                }
-                // CraftBukkit end
+				// Torch - SPIGOT-1132: Duplicate Minecart interact event
 
                 EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.a(world, (double) blockposition.getX() + 0.5D, (double) blockposition.getY() + 0.0625D + d0, (double) blockposition.getZ() + 0.5D, this.b);
 
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
index d412e13..b64399b 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotConfig.java
@@ -32,7 +32,6 @@ public class TacoSpigotConfig {
         CONFIG_FILE = configFile;
         config = new YamlConfiguration();
         try {
-            System.out.println("Loading TacoSpigot configuration from: " + configFile.getName());
             config.load(CONFIG_FILE);
         } catch (IOException ex) {
         } catch (InvalidConfigurationException ex) {
-- 
2.8.2.windows.1

