From 20cde3f06989198cd3f9efeb97a2bcd298b94ef9 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 14:10:26 +0800
Subject: [PATCH] Fix simplify and optimize BlockStateArrays


diff --git a/src/main/java/net/minecraft/server/BlockStateList.java b/src/main/java/net/minecraft/server/BlockStateList.java
index 4c39b86..fdad506 100644
--- a/src/main/java/net/minecraft/server/BlockStateList.java
+++ b/src/main/java/net/minecraft/server/BlockStateList.java
@@ -160,7 +160,8 @@ public class BlockStateList {
             // TacoSpigot start
             this.bAsImmutableMap = immutablemap;
             if (TacoSpigotConfig.useArraysForBlockStates) {
-                b = new ImmutableArrayMap<>(IBlockState.INDEXER, BlockState::getById, immutablemap);
+                ImmutableArrayMap<IBlockState, Comparable> arrayMap = new ImmutableArrayMap<IBlockState, Comparable>(IBlockState::getId, BlockState::getById, (ImmutableMap) immutablemap);
+                b = (key) -> arrayMap.get(key.getId());
             } else {
                 b = immutablemap::get;
             }
-- 
2.8.2.windows.1

