From 42530fb1cb9e28b7c943fd09cfeaa11aa981c125 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 14:35:36 +0800
Subject: [PATCH] Fix DataWatcher for ArrayMap


diff --git a/src/main/java/net/minecraft/server/DataWatcher.java b/src/main/java/net/minecraft/server/DataWatcher.java
index ac7aabe..f94e9d6 100644
--- a/src/main/java/net/minecraft/server/DataWatcher.java
+++ b/src/main/java/net/minecraft/server/DataWatcher.java
@@ -26,7 +26,6 @@ public class DataWatcher {
 
     private static final Map<Class<? extends Entity>, Integer> a = Maps.newHashMap();
     private final Entity b;
-	private final ReadWriteLock d = new ReentrantReadWriteLock();
     private final Int2ObjectMap<Item<?>> c = TacoSpigotConfig.useArraysForDatawatcher ? new ArrayMap<>() : new Int2ObjectOpenHashMap<>(); // Paper // TacoSpigot - use 'Int2ObjectMap' and use 'ArrayMap' if useArraysForDatawatcher
     private final ReadWriteLock d = TacoSpigotConfig.disableDataWatcherReadLocking && TacoSpigotConfig.useArraysForDatawatcher ? NoOpReadWriteLock.INSTANCE : new ReentrantReadWriteLock(); // TacoSpigot - don't lock if we're using ArrayMap
     private boolean e = true;
diff --git a/src/main/java/net/techcable/tacospigot/ArrayMap.java b/src/main/java/net/techcable/tacospigot/ArrayMap.java
index c2d9c8b..48fd7e1 100644
--- a/src/main/java/net/techcable/tacospigot/ArrayMap.java
+++ b/src/main/java/net/techcable/tacospigot/ArrayMap.java
@@ -128,7 +128,7 @@ public final class ArrayMap<V> extends AbstractInt2ObjectMap<V> {
         AtomicReferenceArray<V> values = this.values;
            if (key < 0) {
                throw new IllegalArgumentException(negativeKey(key));
-           } else if (key >= values.length) {
+           } else if (key >= values.length()) {
                return null;
            } else {
                V oldValue = values.getAndSet(key, null);
-- 
2.8.2.windows.1

