From 19b94b6596e4ddb899cc1733a73894b0ed177530 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 14:42:05 +0800
Subject: [PATCH] Simplify getEntity by UUID patch, miscs


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index f065dae..75dd6f9 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1507,6 +1507,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         this.Y = 0L;
     }
 
+	public Entity getEntity(UUID uuid) { return this.a(uuid); } // Paper - OBFHELPER
     @Nullable
     public Entity a(UUID uuid) {
         WorldServer[] aworldserver = this.worldServer;
@@ -1598,19 +1599,4 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
     }
     // CraftBukkit end
 
-    // Paper start
-    public Entity getEntity(UUID uuid) {
-        Entity entity;
-        for (WorldServer world : worldServer) {
-            if (world == null) {
-                continue;
-            }
-            entity = world.getEntity(uuid);
-            if (entity != null && !entity.dead) {
-                return entity;
-            }
-        }
-        return null;
-    }
-    // Paper end
 }
-- 
2.8.2.windows.1

