From 6ad396eae27eabc3849eacdf8d03dc685f5c45eb Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 02:30:13 +0800
Subject: [PATCH] MC-80966 - Always send chunk sections


diff --git a/src/main/java/net/minecraft/server/ChunkSection.java b/src/main/java/net/minecraft/server/ChunkSection.java
index 6cf71b8..55e5e33 100644
--- a/src/main/java/net/minecraft/server/ChunkSection.java
+++ b/src/main/java/net/minecraft/server/ChunkSection.java
@@ -64,7 +64,10 @@ public class ChunkSection {
     }
 
     public boolean a() {
-        return this.nonEmptyBlockCount == 0;
+        // Paper - MC-80966
+		// Torch - backport
+        // Even if there are no blocks, there may be other information associated with the chunk, always send it.
+        return false;
     }
 
     public boolean shouldTick() {
-- 
2.8.2.windows.1

