From 6866f0e2f5055529079d7f577452eef9c8afbc5e Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 15:22:19 +0800
Subject: [PATCH] SPIGOT-661, misc update clients when setting blocks to air


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index cdeff9b..a9666ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -135,6 +135,13 @@ public class CraftBlock implements Block {
     public boolean setTypeIdAndData(final int type, final byte data, final boolean applyPhysics) {
         IBlockData blockData = getNMSBlock(type).fromLegacyData(data);
         BlockPosition position = new BlockPosition(x, y, z);
+		
+		// SPIGOT-611: need to do this to prevent glitchiness. Easier to handle this here (like /setblock) than to fix weirdness in tile entity cleanup
+        // Paper start - If we're setting air, we need to explicitly tell it to send to clients
+        int flag = (type == 0) ? 2 : 0;
+        chunk.getHandle().getWorld().setTypeAndData(position, Blocks.AIR.getBlockData(), flag);
+        // Paper end
+		
         if (applyPhysics) {
             return chunk.getHandle().getWorld().setTypeAndData(position, blockData, 3);
         } else {
-- 
2.8.2.windows.1

