From 975e57a9af4cf72c3495435558fafc14da9ec1cb Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 03:55:01 +0800
Subject: [PATCH] Torch fix SPIGOT-1942 and rebuild


diff --git a/src/main/java/net/minecraft/server/MerchantRecipe.java b/src/main/java/net/minecraft/server/MerchantRecipe.java
index e3a8b1d..175de6e 100644
--- a/src/main/java/net/minecraft/server/MerchantRecipe.java
+++ b/src/main/java/net/minecraft/server/MerchantRecipe.java
@@ -3,7 +3,11 @@ package net.minecraft.server;
 import javax.annotation.Nullable;
 import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe; // CraftBukkit
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
 public class MerchantRecipe {
+	private static final Logger logger = LogManager.getLogger();
 
     public ItemStack buyingItem1;
     public ItemStack buyingItem2;
@@ -120,6 +124,21 @@ public class MerchantRecipe {
 
     public NBTTagCompound k() {
         NBTTagCompound nbttagcompound = new NBTTagCompound();
+		
+		// Torch - SPIGOT-1942
+		if (this.buyingItem1 == null || this.sellingItem == null) {
+           MerchantRecipe.logger.warn("save recipe failed.");
+		   
+           ItemStack buy = new ItemStack(Items.EMERALD, 9);
+           ItemStack sell = new ItemStack(Blocks.EMERALD_BLOCK);
+		   
+           nbttagcompound.set("buy", buy.save(new NBTTagCompound()));
+           nbttagcompound.set("sell", sell.save(new NBTTagCompound()));
+           nbttagcompound.setInt("uses", 1);
+           nbttagcompound.setInt("maxUses", 1);
+           nbttagcompound.setBoolean("rewardExp", false);
+           return nbttagcompound;
+       }
 
         nbttagcompound.set("buy", this.buyingItem1.save(new NBTTagCompound()));
         nbttagcompound.set("sell", this.sellingItem.save(new NBTTagCompound()));
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index c2e099f..1675d16 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -803,7 +803,6 @@ public abstract class World implements IBlockAccess {
             if (previous != null) {
                 return previous;
             }
-        }
         // CraftBukkit end
         Chunk chunk = ((ChunkProviderServer) this.chunkProvider).getChunkIfLoaded(x >> 4, z >> 4);
         if (chunk != null) {
-- 
2.8.2.windows.1

