From 6dc74ba4bf12ab254f1948f6dba6c0830948b34d Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 16:24:20 +0800
Subject: [PATCH] Permit ProtocolSupport to use magic


diff --git a/src/main/java/net/minecraft/server/DataWatcher.java b/src/main/java/net/minecraft/server/DataWatcher.java
index f94e9d6..700d3ed 100644
--- a/src/main/java/net/minecraft/server/DataWatcher.java
+++ b/src/main/java/net/minecraft/server/DataWatcher.java
@@ -81,7 +81,8 @@ public class DataWatcher {
         DataWatcher.Item datawatcher_item = new DataWatcher.Item(datawatcherobject, t0);
 
         this.d.writeLock().lock();
-        this.c.put(datawatcherobject.a(), datawatcher_item); // TacoSpigot - primitives
+		this.c.put(Integer.valueOf(datawatcherobject.a()), datawatcher_item);
+        // this.c.put(datawatcherobject.a(), datawatcher_item); // TacoSpigot - primitives
         this.e = false;
         this.d.writeLock().unlock();
     }
@@ -92,7 +93,8 @@ public class DataWatcher {
         DataWatcher.Item datawatcher_item;
 
         try {
-            datawatcher_item = (DataWatcher.Item) this.c.get(datawatcherobject.a()); // TacoSpigot - use primitives
+			datawatcher_item = (DataWatcher.Item) this.c.get(Integer.valueOf(datawatcherobject.a()));
+            // datawatcher_item = (DataWatcher.Item) this.c.get(datawatcherobject.a()); // TacoSpigot - use primitives
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.a(throwable, "Getting synched entity data");
             CrashReportSystemDetails crashreportsystemdetails = crashreport.a("Synched entity data");
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
index 3f18f79..a82623b 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
@@ -65,7 +65,7 @@ public class TacoSpigotWorldConfig {
 
     public boolean isHopperPushBased;
     private void isHopperPushBased() {
-        isHopperPushBased = getBoolean("hopper.push-based", true);
+        isHopperPushBased = getBoolean("hopper.push-based", true); // Torch - default use
     }
 
     public boolean optimizeArmorStandMovement;
@@ -75,6 +75,6 @@ public class TacoSpigotWorldConfig {
 
     public boolean isHopperFireIMIE;
     private void isHopperFireIMIE() {
-        isHopperFireIMIE = getBoolean("hopper.fire-InventoryMoveItemEvent", true);
+        isHopperFireIMIE = getBoolean("hopper.fire-InventoryMoveItemEvent", false); // Torch - default disable
     }
 }
-- 
2.8.2.windows.1

