From 2e40b6ab556b64b3bf42047574031a766c07c115 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 13:54:26 +0800
Subject: [PATCH] SPIGOT-1599: Allow teleporting disconnected players


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 2c41330..3dd5265 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -533,7 +533,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             return false;
         }
 
-        if (entity.playerConnection == null || entity.playerConnection.isDisconnected()) {
+        if (entity.playerConnection == null) {
            return false;
         }
 
-- 
2.8.2.windows.1

