From 3226a4805bb9912b73e6f90d1721adb3f7d2d8e7 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 04:26:43 +0800
Subject: [PATCH] simple fix mcutil logically


diff --git a/src/main/java/net/minecraft/server/MCUtil.java b/src/main/java/net/minecraft/server/MCUtil.java
index 1e7e298..caa63f6 100644
--- a/src/main/java/net/minecraft/server/MCUtil.java
+++ b/src/main/java/net/minecraft/server/MCUtil.java
@@ -1,5 +1,5 @@
 package net.minecraft.server;
-+
+
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import org.bukkit.Location;
 import org.bukkit.craftbukkit.util.Waitable;
-- 
2.8.2.windows.1

