From 8278a5d8912a817001fbcb77b5e95ca0824715ab Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 1 Jul 2016 17:30:00 +0800
Subject: [PATCH] Permit ProtocolSupport to use magic actually


diff --git a/src/main/java/net/minecraft/server/PacketDataSerializer.java b/src/main/java/net/minecraft/server/PacketDataSerializer.java
index cad4694..62f8a2d 100644
--- a/src/main/java/net/minecraft/server/PacketDataSerializer.java
+++ b/src/main/java/net/minecraft/server/PacketDataSerializer.java
@@ -237,7 +237,7 @@ public class PacketDataSerializer extends ByteBuf {
 
     public PacketDataSerializer a(@Nullable ItemStack itemstack) {
 		// TacoSpigot start - fire ItemSendEvent
-        if (player == null) throw new UnsupportedOperationException("Can't write an ItemStack when a serializer doesn't know the player");
+        // if (player == null) throw new UnsupportedOperationException("Can't write an ItemStack when a serializer doesn't know the player"); // Torch - Permit ProtocolSupport to use magic
         ItemSendEvent event = new ItemSendEvent(player.getBukkitEntity(), itemstack != null ? CraftItemStack.asBukkitCopy(itemstack) : null);
         event.callEvent();
         itemstack = CraftItemStack.asNMSCopy(event.getStack());
-- 
2.8.2.windows.1

