From 255d7a02bd858860731bbe7ee18acc4f7e970bc8 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 16:54:14 +0800
Subject: [PATCH] Fix exception that shouldnt throw


diff --git a/src/main/java/net/minecraft/server/EntityTracker.java b/src/main/java/net/minecraft/server/EntityTracker.java
index 0973c50..9002973 100644
--- a/src/main/java/net/minecraft/server/EntityTracker.java
+++ b/src/main/java/net/minecraft/server/EntityTracker.java
@@ -141,11 +141,13 @@ public class EntityTracker {
             entity.appendEntityCrashDetails(crashreportsystemdetails);
             ((EntityTrackerEntry) this.trackedEntities.get(entity.getId())).b().appendEntityCrashDetails(crashreport.a("Entity That Is Already Tracked"));
 
+			/*
             try {
                 throw new ReportedException(crashreport);
             } catch (ReportedException reportedexception) {
                 EntityTracker.a.error("\"Silently\" catching entity tracking error.", reportedexception);
             }
+			*/
         }
 
     }
-- 
2.8.2.windows.1

