From a586c0d557fb802921e9d7ee01b5aeea0c2332fd Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 07:26:48 +0800
Subject: [PATCH] Add Player.stopSound, miscs


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 4f1a2e3..f90b1d4 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2230,6 +2230,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                 itemstack1 = this.player.getItemInMainHand();
                 if (itemstack1 != null) {
                     if (itemstack.getItem() == Items.WRITABLE_BOOK && itemstack.getItem() == itemstack1.getItem()) {
+						itemstack1 = new ItemStack(Items.WRITABLE_BOOK); // CraftBukkit
                         itemstack1.a("pages", (NBTBase) itemstack.getTag().getList("pages", 8));
                         CraftEventFactory.handleEditBookEvent(player, itemstack1); // Paper
                     }
@@ -2302,6 +2303,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     }
                 } catch (Exception exception2) {
                     PlayerConnection.LOGGER.error("Couldn\'t select trade", exception2);
+					this.disconnect("Invalid trade data!"); // CraftBukkit
                 }
             } else {
                 TileEntity tileentity;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index a9654be..bac36e7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -404,6 +404,21 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         PacketPlayOutCustomSoundEffect packet = new PacketPlayOutCustomSoundEffect(sound, SoundCategory.MASTER, loc.getX(), loc.getY(), loc.getZ(), volume, pitch);
         getHandle().playerConnection.sendPacket(packet);
     }
+	
+	@Override
+    public void stopSound(Sound sound) {
+        stopSound(CraftSound.getSound(sound));
+    }
+
+    @Override
+    public void stopSound(String sound) {
+        if (getHandle().playerConnection == null) return;
+        PacketDataSerializer packetdataserializer = new PacketDataSerializer(Unpooled.buffer());
+
+        packetdataserializer.a(sound);
+        packetdataserializer.a("");
+        getHandle().playerConnection.sendPacket(new PacketPlayOutCustomPayload("MC|StopSound", packetdataserializer));
+    }
 
     @Override
     public void playEffect(Location loc, Effect effect, int data) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
index bfa00ec..0329c26 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
@@ -44,7 +44,7 @@ public class CraftVillager extends CraftAgeable implements Villager, InventoryHo
 
     public void setProfession(Profession profession) {
 		Validate.isTrue(0 < profession.ordinal() && profession.ordinal() < Profession.HUSK.ordinal(), "This profession is reserved for Zombies: ", profession);
-        getHandle().setProfession(profession.ordinal() - 1);
+        getHandle().setProfession(profession.getId() - 1);
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java
index 35b5480..80db092 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftZombie.java
@@ -59,6 +59,6 @@ public class CraftZombie extends CraftMonster implements Zombie {
     @Override
     public Villager.Profession getVillagerProfession() {
         // if (!isVillager()) return null; // Torch - better map Zombie/Villager professions.
-        return Villager.Profession.values()[getHandle().getVillagerType().ordinal()];
+        return Villager.Profession.getProfession(getHandle().getVillagerType());
     }
 }
-- 
2.8.2.windows.1

