From e17291e007202dfd8a60ef6826fde752317bd059 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 09:00:10 +0800
Subject: [PATCH] simple fix chunk get


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 0c87f4b..25fe7d1 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -796,11 +796,14 @@ public abstract class World implements IBlockAccess {
 
     // Paper start - reduces need to do isLoaded before getType
     public IBlockData getTypeIfLoaded(BlockPosition blockposition) {
+        final int x = blockposition.getX();
+        final int y = blockposition.getY();
+        final int z = blockposition.getZ();
         if (captureTreeGeneration) return findCaptured(blockposition); // use seperate findCaptured() method
-        Chunk chunk = this.getChunkIfLoaded(blockposition);
+        Chunk chunk = ((ChunkProviderServer) this.chunkProvider).getChunkIfLoaded(x >> 4, z >> 4);
         if (chunk != null) {
-            return chunk.getBlockData(x, y, z);
-        }
+           final IBlockData previous = getCapturedBlockType(x, y, z);
+		}
         return null;
     }
     // Paper end
-- 
2.8.2.windows.1

