From 8bfa0ff803cb5b037b68ba535e818e2f8fc66993 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 09:13:13 +0800
Subject: [PATCH] fix api profile simple


diff --git a/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java b/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java
index c9e6c50..f6d6b88 100644
--- a/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java
+++ b/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java
@@ -8,6 +8,7 @@ import com.google.common.base.Preconditions;
 import org.bukkit.Bukkit;
 import org.bukkit.OfflinePlayer;
 import org.bukkit.entity.Player;
+import org.bukkit.craftbukkit.CraftServer;
 
 /**
 * Represents a player's profile
@@ -140,7 +141,7 @@ public final class AccountProfile {
     * @throws LookupFailedException    if unable to lookup properties
     */
    public AccountProfile lookupProperties() {
-       return withProperties(Bukkit.getProfileLookup().lookupProperties(this));
+       return withProperties(CraftServer.getProfileLookup().lookupProperties(this));
    }
 
    /**
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 972af98..6249394 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -9,6 +9,7 @@ import java.io.InputStreamReader;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
+import com.destroystokyo.paper.profile.AccountProfile; // Paper
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashMap;
@@ -1892,4 +1893,6 @@ public final class CraftServer implements Server {
         return entity.getBukkitEntity();
     }
     // Paper end
+	
+	public ProfileLookup getProfileLookup();
 }
-- 
2.8.2.windows.1

