From db7847a896d9fa0591593e66684d4e7666a93925 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 07:59:53 +0800
Subject: [PATCH] SPIGOT-1224 misc SPIGOT-1284, miscs


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 3bb2d73..b2e7174 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -63,7 +63,6 @@ public class Chunk {
 
     // CraftBukkit start - Neighbor loaded cache for chunk lighting and entity ticking
     private int neighbors = 0x1 << 12;
-	public long chunkKey;
 
     public boolean areNeighborsLoaded(final int radius) {
         switch (radius) {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 75dd6f9..6779069 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -776,6 +776,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
 
     public void D() {
         MinecraftTimings.minecraftSchedulerTimer.startTiming(); // Paper
+		this.server.getScheduler().mainThreadHeartbeat(this.ticks); // CraftBukkit
         this.methodProfiler.a("jobs");
         Queue queue = this.j;
 
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalTempt.java b/src/main/java/net/minecraft/server/PathfinderGoalTempt.java
index a04a4b8..6fb99d9 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalTempt.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalTempt.java
@@ -45,9 +45,9 @@ public class PathfinderGoalTempt extends PathfinderGoal {
             return false;
         } else {
             this.h = this.a.world.findNearbyPlayer(this.a, 10.0D);
-             // CraftBukkit start
+            // CraftBukkit start
             // PAIL: rename
-            boolean tempt = this.h == null ? false : this.a(this.h.getItemInMainHand()) || this.a(this.h.getItemInOffHand());
+			boolean tempt = this.h == null ? false : this.a(this.h.getItemInMainHand()) || this.a(this.h.getItemInOffHand());
             if (tempt) {
                 EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(this.a, this.h, EntityTargetEvent.TargetReason.TEMPT);
                 if (event.isCancelled()) {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index f90b1d4..24b8145 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1902,7 +1902,11 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     }
 
                     event.setCancelled(cancelled);
+					Container oldContainer = this.player.activeContainer; // SPIGOT-1224
                     server.getPluginManager().callEvent(event);
+					if (this.player.activeContainer != oldContainer) {
+                        return;
+                    }
 
                     switch (event.getResult()) {
                         case ALLOW:
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
index 16fe364..3322f24 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
@@ -43,7 +43,7 @@ public class CraftVillager extends CraftAgeable implements Villager, InventoryHo
     }
 
     public void setProfession(Profession profession) {
-		Validate.isTrue(0 < profession.getId() && profession.getId() < Profession.HUSK.getId(), "This profession is reserved for Zombies: ", profession);
+		Validate.notNull(profession); // Torch
         getHandle().setProfession(profession.getId() - 1);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryDoubleChest.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryDoubleChest.java
index f542c07..7b07239 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryDoubleChest.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryDoubleChest.java
@@ -7,6 +7,8 @@ import org.bukkit.inventory.DoubleChestInventory;
 import org.bukkit.inventory.Inventory;
 import org.bukkit.inventory.ItemStack;
 
+import org.bukkit.Location; // Torch
+
 import net.minecraft.server.InventoryLargeChest;
 
 public class CraftInventoryDoubleChest extends CraftInventory implements DoubleChestInventory {
-- 
2.8.2.windows.1

