From 44734e81c979c86e072b88304949e4aac4433bd7 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 09:18:05 +0800
Subject: [PATCH] simple fix repeating


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6249394..134b90d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1894,5 +1894,4 @@ public final class CraftServer implements Server {
     }
     // Paper end
 	
-	public ProfileLookup getProfileLookup();
 }
-- 
2.8.2.windows.1

