From 443554166652199bcbbc37661d47c9e159a5887b Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 08:07:24 +0800
Subject: [PATCH] SPIGOT-1364: sendMap should update entire map, miscs


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index b2e7174..c5265e2 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -16,7 +16,7 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 import com.google.common.collect.Lists; // CraftBukkit
-import org.bukkit.Bukkit; // CraftBukkit
+import org.bukkit.Server; // CraftBukkit
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 
 import java.util.Queue;
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalTempt.java b/src/main/java/net/minecraft/server/PathfinderGoalTempt.java
index 6fb99d9..773c915 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalTempt.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalTempt.java
@@ -11,7 +11,7 @@ import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
 
 public class PathfinderGoalTempt extends PathfinderGoal {
 
-    private EntityCreature a;
+    private Entity a;
     private double b;
     private double c;
     private double d;
@@ -24,11 +24,11 @@ public class PathfinderGoalTempt extends PathfinderGoal {
     private Set<Item> k;
     private boolean l;
 
-    public PathfinderGoalTempt(EntityCreature entitycreature, double d0, Item item, boolean flag) {
+    public PathfinderGoalTempt(Entity entitycreature, double d0, Item item, boolean flag) {
         this(entitycreature, d0, flag, Sets.newHashSet(new Item[] { item}));
     }
 
-    public PathfinderGoalTempt(EntityCreature entitycreature, double d0, boolean flag, Set<Item> set) {
+    public PathfinderGoalTempt(Entity entitycreature, double d0, boolean flag, Set<Item> set) {
         this.a = entitycreature;
         this.b = d0;
         this.k = set;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 12bf4e6..2c41330 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -521,7 +521,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             }
         }
 
-        PacketPlayOutMap packet = new PacketPlayOutMap(map.getId(), map.getScale().getValue(), true, icons, data.buffer, 0, 0, 0, 0);
+        PacketPlayOutMap packet = new PacketPlayOutMap(map.getId(), map.getScale().getValue(), true, icons, data.buffer, 0, 0, 128, 128);
         getHandle().playerConnection.sendPacket(packet);
     }
 
-- 
2.8.2.windows.1

