From f4a27c1785b2301e4bfd6e0d179134cd6e0e7ba5 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 08:17:41 +0800
Subject: [PATCH] Fix chunk missed, revert SPIGOT-1468: ENDER_PEARL SpawnReason


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index c5265e2..3b724db 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -17,6 +17,7 @@ import org.apache.logging.log4j.Logger;
 
 import com.google.common.collect.Lists; // CraftBukkit
 import org.bukkit.Server; // CraftBukkit
+import org.bukkit.Bukkit; // Torch
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 
 import java.util.Queue;
diff --git a/src/main/java/net/minecraft/server/EntityEnderPearl.java b/src/main/java/net/minecraft/server/EntityEnderPearl.java
index 4db52ec..012bb53 100644
--- a/src/main/java/net/minecraft/server/EntityEnderPearl.java
+++ b/src/main/java/net/minecraft/server/EntityEnderPearl.java
@@ -73,7 +73,7 @@ public class EntityEnderPearl extends EntityProjectile {
 
                             entityendermite.a(true);
                             entityendermite.setPositionRotation(entityliving.locX, entityliving.locY, entityliving.locZ, entityliving.yaw, entityliving.pitch);
-                            this.world.addEntity(entityendermite, CreatureSpawnEvent.SpawnReason.ENDER_PEARL);
+                            this.world.addEntity(entityendermite);
                         }
 
                         if (entityliving.isPassenger()) {
-- 
2.8.2.windows.1

