From e1b43d78309dd53917e5564ca6240e20d4f3e363 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 26 Jun 2016 09:47:46 +0800
Subject: [PATCH] automate patch


diff --git a/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java b/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java
index 532e549..fc11c05 100644
--- a/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java
+++ b/src/main/java/com/destroystokyo/paper/profile/AccountProfile.java
@@ -141,7 +141,7 @@ public final class AccountProfile {
     * @throws LookupFailedException    if unable to lookup properties
     */
    public AccountProfile lookupProperties() {
-       return withProperties(Bukkit.getServer().lookupProperties(this)); // Torch - skipped
+       return withProperties(MinecraftServer.getServer().lookupProperties(this)); // Torch - skipped
    }
 
    /**
-- 
2.8.2.windows.1

