From 42912f1cfc03a1926c7e4e3f14c9f25a68dc0f54 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 17:27:27 +0800
Subject: [PATCH] SPIGOT-2504: Save structure info for secondary worlds


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 11c4913..298406e 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -1106,10 +1106,6 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         this.worldData.k(this.getWorldBorder().getWarningTime());
         this.worldData.b(this.getWorldBorder().j());
         this.worldData.e(this.getWorldBorder().i());
-        // CraftBukkit start - save worldMaps once, rather than once per shared world
-        if (!(this instanceof SecondaryWorldServer)) {
-            this.worldMaps.a();
-        }
         this.dataManager.saveWorldData(this.worldData, this.server.getPlayerList().t());
         // CraftBukkit end
     }
-- 
2.8.2.windows.1

