From 0deeaf8fd655eb91dc04645fa612db249b9fb68b Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 17:32:54 +0800
Subject: [PATCH] SPIGOT-1994: Revert collisions to Vanilla behaviour


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 816de2d..81092bd 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -2024,13 +2024,6 @@ public abstract class EntityLiving extends Entity {
             Iterator it = list.iterator();
             while (it.hasNext()) {
                 Entity entity = (Entity) it.next();
-                // TODO better check now?
-                // CraftBukkit start - Only handle mob (non-player) collisions every other tick
-                if (entity instanceof EntityLiving && !(this instanceof EntityPlayer) && this.ticksLived % 2 == 0) {
-                    continue;
-                }
-                // CraftBukkit end
-
                 entity.numCollisions++; // Spigot
                 numCollisions++; // Spigot
                 this.C(entity);
-- 
2.8.2.windows.1

