From 3b1c4f84306c9f3318a78944a143ad0aa87b57c5 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 12:46:22 +0800
Subject: [PATCH] Fix thread, misc SPIGOT-2405 and SPIGOT-2408


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 3375cfa..64e8b09 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -1161,7 +1161,7 @@ public abstract class Entity implements ICommandListener {
                     ++this.fireTicks;
                     // CraftBukkit start - Not on fire yet
                     if (this.fireTicks <= 0) { // Only throw events on the first combust, otherwise it spams
-                        EntityCombustEvent event = new EntityCombustEvent(getBukkitEntity(), 8);
+                        EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(null, getBukkitEntity(), 8); // Torch - SPIGOT-2405
                         world.getServer().getPluginManager().callEvent(event);
 
                         if (!event.isCancelled()) {
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 733e5eb..84f1190 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1533,7 +1533,7 @@ public abstract class World implements IBlockAccess {
 				this.getChunkAt(k, j).b(entity);
             }
 			
-			this.c(entity)
+			this.c(entity);
         }
 		// Torch end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingSand.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingSand.java
index 65e1408..2471a0b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingSand.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFallingSand.java
@@ -64,4 +64,12 @@ public class CraftFallingSand extends CraftEntity implements FallingSand {
         return getOrigin();
     }
     // Paper end
+	
+	@Override
+    public void setTicksLived(int value) {
+        super.setTicksLived(value);
+
+        // Second field for EntityFallingBlock
+        getHandle().ticksLived = value;
+    }
 }
-- 
2.8.2.windows.1

