From 1aa2bea2cd9416d117c433cc39e4f92c0f81ce9a Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 12:51:43 +0800
Subject: [PATCH] Fix threaded actually


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 84f1190..5541fe7 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1526,8 +1526,8 @@ public abstract class World implements IBlockAccess {
         while (it.hasNext()) {
             entity = (Entity) it.next();
 			
-            j = e.getChunkX();
-			k = e.getChunkZ();
+            j = entity.getChunkX();
+			k = entity.getChunkZ();
 			
             if (entity.aa && this.isChunkLoaded(j, k, true)) {
 				this.getChunkAt(k, j).b(entity);
-- 
2.8.2.windows.1

