From b0decf96631f2726a8344ce4dc7440c6419a170d Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 12:55:19 +0800
Subject: [PATCH] Fix multi threaded this time


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index e500764..3627ac0 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -33,6 +33,7 @@ public class ChunkProviderServer implements IChunkProvider {
     private final IChunkLoader chunkLoader;
     // Paper start
     protected Chunk lastChunkByPos = null;
+	/*
     public Long2ObjectOpenHashMap<Chunk> chunks = new Long2ObjectOpenHashMap<Chunk>(8192) {
 
         @Override
@@ -51,6 +52,7 @@ public class ChunkProviderServer implements IChunkProvider {
             return super.remove(key);
         }
     }; // CraftBukkit
+	*/
     // Paper end
     public final WorldServer world;
 
-- 
2.8.2.windows.1

