From cca0ba0aa66fe38890d2ea54c8933d9a30abd661 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 17:45:55 +0800
Subject: [PATCH] SPIGOT-2490: Villager farming event


diff --git a/src/main/java/net/minecraft/server/BlockRedstoneOre.java b/src/main/java/net/minecraft/server/BlockRedstoneOre.java
index ccccb7b..3117e58 100644
--- a/src/main/java/net/minecraft/server/BlockRedstoneOre.java
+++ b/src/main/java/net/minecraft/server/BlockRedstoneOre.java
@@ -61,7 +61,7 @@ public class BlockRedstoneOre extends Block {
         this.playEffect(world, blockposition);
         if (this == Blocks.REDSTONE_ORE) {
             // CraftBukkit start
-            if (CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.LIT_REDSTONE_ORE, 0).isCancelled()) {
+            if (CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, Blocks.LIT_REDSTONE_ORE, 0).isCancelled()) {
                 return;
             }
             // CraftBukkit end
diff --git a/src/main/java/net/minecraft/server/BlockSoil.java b/src/main/java/net/minecraft/server/BlockSoil.java
index 64394b5..d228ac9 100644
--- a/src/main/java/net/minecraft/server/BlockSoil.java
+++ b/src/main/java/net/minecraft/server/BlockSoil.java
@@ -76,7 +76,7 @@ public class BlockSoil extends Block {
                 return;
             }
 
-            if (CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.DIRT, 0).isCancelled()) {
+            if (CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, Blocks.DIRT, 0).isCancelled()) {
                 return;
             }
             // CraftBukkit end
diff --git a/src/main/java/net/minecraft/server/BlockWaterLily.java b/src/main/java/net/minecraft/server/BlockWaterLily.java
index 4805627..09b8c41 100644
--- a/src/main/java/net/minecraft/server/BlockWaterLily.java
+++ b/src/main/java/net/minecraft/server/BlockWaterLily.java
@@ -20,7 +20,7 @@ public class BlockWaterLily extends BlockPlant {
 
     public void a(World world, BlockPosition blockposition, IBlockData iblockdata, Entity entity) {
         super.a(world, blockposition, iblockdata, entity);
-        if (entity instanceof EntityBoat && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) { // CraftBukkit
+        if (entity instanceof EntityBoat && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, Blocks.AIR, 0).isCancelled()) { // CraftBukkit
             world.setAir(new BlockPosition(blockposition), true);
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityEnderman.java b/src/main/java/net/minecraft/server/EntityEnderman.java
index bfa6621..2c36dc0 100644
--- a/src/main/java/net/minecraft/server/EntityEnderman.java
+++ b/src/main/java/net/minecraft/server/EntityEnderman.java
@@ -358,7 +358,7 @@ public class EntityEnderman extends EntityMonster {
 
             if (iblockdata2 != null && this.a(world, blockposition, iblockdata2.getBlock(), iblockdata, iblockdata1)) {
                 // CraftBukkit start - Place event
-                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.a, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this.a.getCarried().getBlock(), this.a.getCarried().getBlock().toLegacyData(this.a.getCarried())).isCancelled()) {
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.a, blockposition, this.a.getCarried().getBlock(), this.a.getCarried().getBlock().toLegacyData(this.a.getCarried())).isCancelled()) {
                 world.setTypeAndData(blockposition, iblockdata2, 3);
                 this.a.setCarried((IBlockData) null);
                 }
diff --git a/src/main/java/net/minecraft/server/EntityFallingBlock.java b/src/main/java/net/minecraft/server/EntityFallingBlock.java
index 8ae5606..10592bf 100644
--- a/src/main/java/net/minecraft/server/EntityFallingBlock.java
+++ b/src/main/java/net/minecraft/server/EntityFallingBlock.java
@@ -69,7 +69,7 @@ public class EntityFallingBlock extends Entity {
 
             if (this.ticksLived++ == 0) {
                 blockposition = new BlockPosition(this);
-                if (this.world.getType(blockposition).getBlock() == block && !CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) {
+                if (this.world.getType(blockposition).getBlock() == block && !CraftEventFactory.callEntityChangeBlockEvent(this, blockposition, Blocks.AIR, 0).isCancelled()) {
                     this.world.setAir(blockposition);
 					this.world.paperConfig.antiXrayInstance.updateNearbyBlocks(this.world, blockposition); // Paper - Anti-Xray
                 } else if (!this.world.isClientSide) {
@@ -111,7 +111,7 @@ public class EntityFallingBlock extends Entity {
                         if (!this.f) {
                             // CraftBukkit start
                             if (this.world.a(block, blockposition, true, EnumDirection.UP, (Entity) null, (ItemStack) null) && !BlockFalling.i(this.world.getType(blockposition.down()))) {
-                                if (CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), this.block.getBlock(), this.block.getBlock().toLegacyData(this.block)).isCancelled()) {
+                                if (CraftEventFactory.callEntityChangeBlockEvent(this, blockposition, this.block.getBlock(), this.block.getBlock().toLegacyData(this.block)).isCancelled()) {
                                     return;
                                 }
                                 this.world.setTypeAndData(blockposition, this.block, 3);
diff --git a/src/main/java/net/minecraft/server/EntityPotion.java b/src/main/java/net/minecraft/server/EntityPotion.java
index c70c9b7..f04857c 100644
--- a/src/main/java/net/minecraft/server/EntityPotion.java
+++ b/src/main/java/net/minecraft/server/EntityPotion.java
@@ -194,7 +194,7 @@ public class EntityPotion extends EntityProjectile {
 
     private void a(BlockPosition blockposition) {
         if (this.world.getType(blockposition).getBlock() == Blocks.FIRE) {
-            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) return; // CraftBukkit
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockposition, Blocks.AIR, 0).isCancelled()) return; // CraftBukkit
             this.world.setTypeAndData(blockposition, Blocks.AIR.getBlockData(), 2);
         }
 
diff --git a/src/main/java/net/minecraft/server/EntityRabbit.java b/src/main/java/net/minecraft/server/EntityRabbit.java
index 856d8f6..7cffe67 100644
--- a/src/main/java/net/minecraft/server/EntityRabbit.java
+++ b/src/main/java/net/minecraft/server/EntityRabbit.java
@@ -415,7 +415,7 @@ public class EntityRabbit extends EntityAnimal {
 
                     if (integer.intValue() == 0) {
 						// CraftBukkit start
-                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.c, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.AIR, 0).isCancelled()) {
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.c, blockposition, Blocks.AIR, 0).isCancelled()) {
                             return;
                         }
                         // CraftBukkit end
@@ -425,7 +425,7 @@ public class EntityRabbit extends EntityAnimal {
 						// CraftBukkit start
                         if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(
                                 this.c,
-                                blockposition.getX(), blockposition.getY(), blockposition.getZ(),
+                                blockposition,
                                 block, block.toLegacyData(iblockdata.set(BlockCarrots.AGE, Integer.valueOf(integer.intValue() - 1)))
                         ).isCancelled()) {
                             return;
diff --git a/src/main/java/net/minecraft/server/EntitySilverfish.java b/src/main/java/net/minecraft/server/EntitySilverfish.java
index ed8a425..63c6bb6 100644
--- a/src/main/java/net/minecraft/server/EntitySilverfish.java
+++ b/src/main/java/net/minecraft/server/EntitySilverfish.java
@@ -152,7 +152,7 @@ public class EntitySilverfish extends EntityMonster {
 
                 if (BlockMonsterEggs.i(iblockdata)) {
                     // CraftBukkit start
-                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockposition.getX(), blockposition.getY(), blockposition.getZ(), Blocks.MONSTER_EGG, Block.getId(BlockMonsterEggs.getById(iblockdata.getBlock().toLegacyData(iblockdata)))).isCancelled()) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockposition, Blocks.MONSTER_EGG, Block.getId(BlockMonsterEggs.getById(iblockdata.getBlock().toLegacyData(iblockdata)))).isCancelled()) {
                         return;
                     }
                     // CraftBukkit end
@@ -200,7 +200,7 @@ public class EntitySilverfish extends EntityMonster {
 
                             if (iblockdata.getBlock() == Blocks.MONSTER_EGG) {
                                 // CraftBukkit start
-                                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockposition1.getX(), blockposition1.getY(), blockposition1.getZ(), Blocks.AIR, 0).isCancelled()) {
+                                 if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockposition1, Blocks.AIR, 0).isCancelled()) {
                                     continue;
                                 }
                                 // CraftBukkit end
diff --git a/src/main/java/net/minecraft/server/EntityWither.java b/src/main/java/net/minecraft/server/EntityWither.java
index af86437..2fc52f8 100644
--- a/src/main/java/net/minecraft/server/EntityWither.java
+++ b/src/main/java/net/minecraft/server/EntityWither.java
@@ -314,7 +314,7 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
 
                                 if (iblockdata.getMaterial() != Material.AIR && a(block)) {
                                     // CraftBukkit start
-                                    if (CraftEventFactory.callEntityChangeBlockEvent(this, j2, k2, l2, Blocks.AIR, 0).isCancelled()) {
+                                    if (CraftEventFactory.callEntityChangeBlockEvent(this, blockposition, Blocks.AIR, 0).isCancelled()) {
                                         continue;
                                     }
                                     // CraftBukkit end
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalVillagerFarm.java b/src/main/java/net/minecraft/server/PathfinderGoalVillagerFarm.java
new file mode 100644
index 0000000..2e67050
--- /dev/null
+++ b/src/main/java/net/minecraft/server/PathfinderGoalVillagerFarm.java
@@ -0,0 +1,125 @@
+package net.minecraft.server;
+
+public class PathfinderGoalVillagerFarm extends PathfinderGoalGotoTarget {
+
+    private final EntityVillager c;
+    private boolean d;
+    private boolean e;
+    private int f;
+
+    public PathfinderGoalVillagerFarm(EntityVillager entityvillager, double d0) {
+        super(entityvillager, d0, 16);
+        this.c = entityvillager;
+    }
+
+    public boolean a() {
+        if (this.a <= 0) {
+            if (!this.c.world.getGameRules().getBoolean("mobGriefing")) {
+                return false;
+            }
+
+            this.f = -1;
+            this.d = this.c.dj();
+            this.e = this.c.di();
+        }
+
+        return super.a();
+    }
+
+    public boolean b() {
+        return this.f >= 0 && super.b();
+    }
+
+    public void c() {
+        super.c();
+    }
+
+    public void d() {
+        super.d();
+    }
+
+    public void e() {
+        super.e();
+        this.c.getControllerLook().a((double) this.b.getX() + 0.5D, (double) (this.b.getY() + 1), (double) this.b.getZ() + 0.5D, 10.0F, (float) this.c.N());
+        if (this.f()) {
+            World world = this.c.world;
+            BlockPosition blockposition = this.b.up();
+            IBlockData iblockdata = world.getType(blockposition);
+            Block block = iblockdata.getBlock();
+
+            if (this.f == 0 && block instanceof BlockCrops && ((BlockCrops) block).y(iblockdata)) {
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.c, blockposition, Blocks.AIR, 0).isCancelled()) {
+                    world.setAir(blockposition, true);
+                }
+                // CraftBukkit end
+            } else if (this.f == 1 && block == Blocks.AIR) {
+                InventorySubcontainer inventorysubcontainer = this.c.df();
+
+                for (int i = 0; i < inventorysubcontainer.getSize(); ++i) {
+                    ItemStack itemstack = inventorysubcontainer.getItem(i);
+                    boolean flag = false;
+
+                    if (itemstack != null) {
+						// CraftBukkit start
+                        Block planted = null;
+                        if (itemstack.getItem() == Items.WHEAT_SEEDS) {
+                            planted = Blocks.WHEAT;
+                            flag = true;
+                        } else if (itemstack.getItem() == Items.POTATO) {
+                            planted = Blocks.POTATOES;
+                            flag = true;
+                        } else if (itemstack.getItem() == Items.CARROT) {
+                            planted = Blocks.CARROTS;
+                            flag = true;
+                        } else if (itemstack.getItem() == Items.BEETROOT_SEEDS) {
+                            planted = Blocks.BEETROOT;
+                            flag = true;
+                        }
+						
+						 if (planted != null && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.c, blockposition, planted, 0).isCancelled()) {
+                            world.setTypeAndData(blockposition, planted.getBlockData(), 3);
+                        } else {
+                            flag = false;
+                        }
+                        // CraftBukkit end
+                    }
+
+                    if (flag) {
+                        --itemstack.count;
+                        if (itemstack.count <= 0) {
+                            inventorysubcontainer.setItem(i, (ItemStack) null);
+                        }
+                        break;
+                    }
+                }
+            }
+
+            this.f = -1;
+            this.a = 10;
+        }
+
+    }
+
+    protected boolean a(World world, BlockPosition blockposition) {
+        Block block = world.getType(blockposition).getBlock();
+
+        if (block == Blocks.FARMLAND) {
+            blockposition = blockposition.up();
+            IBlockData iblockdata = world.getType(blockposition);
+
+            block = iblockdata.getBlock();
+            if (block instanceof BlockCrops && ((BlockCrops) block).y(iblockdata) && this.e && (this.f == 0 || this.f < 0)) {
+                this.f = 0;
+                return true;
+            }
+
+            if (block == Blocks.AIR && this.d && (this.f == 1 || this.f < 0)) {
+                this.f = 1;
+                return true;
+            }
+        }
+
+        return false;
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 12bea04..613cf57 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -684,8 +684,8 @@ public class CraftEventFactory {
         return callEntityChangeBlockEvent(entity.getBukkitEntity(), block, material, 0, cancelled);
     }
 
-    public static EntityChangeBlockEvent callEntityChangeBlockEvent(Entity entity, int x, int y, int z, net.minecraft.server.Block type, int data) {
-        Block block = entity.world.getWorld().getBlockAt(x, y, z);
+    public static EntityChangeBlockEvent callEntityChangeBlockEvent(Entity entity, BlockPosition position, net.minecraft.server.Block type, int data) {
+        Block block = entity.world.getWorld().getBlockAt(position.getX(), position.getY(), position.getZ());
         Material material = CraftMagicNumbers.getMaterial(type);
 
         return callEntityChangeBlockEvent(entity.getBukkitEntity(), block, material, data);
-- 
2.8.2.windows.1

