From 48199fbe4d68ef2743b6f5214efa45f6b073d4e7 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 14:30:35 +0800
Subject: [PATCH] Void default move checking


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 24b8145..aac7b15 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -339,13 +339,15 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                 }
 
                 d8 = d5 - entity.locZ;
-                d10 = d6 * d6 + d7 * d7 + d8 * d8;
+                // d10 = d6 * d6 + d7 * d7 + d8 * d8;
                 boolean flag1 = false;
 
+				/*
                 if (d10 > org.spigotmc.SpigotConfig.movedWronglyThreshold) { // Spigot
                     flag1 = true;
                     PlayerConnection.LOGGER.warn(entity.getName() + " moved wrongly!");
                 }
+				*/
 
                 entity.setLocation(d3, d4, d5, f, f1);
                 boolean flag2 = worldserver.getCubes(entity, entity.getBoundingBox().shrink(0.0625D)).isEmpty();
@@ -556,13 +558,15 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                         }
 
                         d9 = d6 - this.player.locZ;
-                        d11 = d7 * d7 + d8 * d8 + d9 * d9;
+                        // d11 = d7 * d7 + d8 * d8 + d9 * d9;
                         boolean flag1 = false;
 
+						/*
                         if (!this.player.K() && d11 > org.spigotmc.SpigotConfig.movedWronglyThreshold && !this.player.isSleeping() && !this.player.playerInteractManager.isCreative() && this.player.playerInteractManager.getGameMode() != WorldSettings.EnumGamemode.SPECTATOR) { // Spigot
                             flag1 = true;
                             PlayerConnection.LOGGER.warn(this.player.getName() + " moved wrongly!");
                         }
+						*/
 
                         this.player.setLocation(d4, d5, d6, f, f1);
                         this.player.checkMovement(this.player.locX - d0, this.player.locY - d1, this.player.locZ - d2);
-- 
2.8.2.windows.1

