From 39561a9aa5c4d292225c551e5572eabe16cba106 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 16:20:53 +0800
Subject: [PATCH] Fix go deeper multi threaded


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 200dbf7..78ccaa2 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -406,7 +406,7 @@ public abstract class EntityHuman extends EntityLiving {
                     this.c(entity);
                 }
             }
-
+		}
     }
 
     private void c(Entity entity) {
-- 
2.8.2.windows.1

