From f2bafc35638455e513efeecdf58322bd6727dace Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 16:31:16 +0800
Subject: [PATCH] Fix import for multi threaded actually


diff --git a/src/main/java/net/minecraft/server/EntityBoat.java b/src/main/java/net/minecraft/server/EntityBoat.java
index c3c7aea..bd14f1f 100644
--- a/src/main/java/net/minecraft/server/EntityBoat.java
+++ b/src/main/java/net/minecraft/server/EntityBoat.java
@@ -5,6 +5,9 @@ import java.util.ArrayList;
 import java.util.List;
 import javax.annotation.Nullable;
 
+import java.util.Iterator;
+import java.util.Queue;
+
 // CraftBukkit start
 import org.bukkit.Location;
 import org.bukkit.entity.Vehicle;
-- 
2.8.2.windows.1

