From a54bd6d99bb39e9b46843de4df29b5ef637b4901 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 21:54:27 +0800
Subject: [PATCH] Add gc command


diff --git a/src/main/java/org/bukkit/command/defaults/GCCommand.java b/src/main/java/org/bukkit/command/defaults/GCCommand.java
new file mode 100644
index 0000000..0bd448d
--- /dev/null
+++ b/src/main/java/org/bukkit/command/defaults/GCCommand.java
@@ -0,0 +1,43 @@
+package org.bukkit.command.defaults;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.lang3.Validate;
+import org.bukkit.Bukkit;
+import org.bukkit.ChatColor;
+import org.bukkit.Material;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+import org.bukkit.entity.Player;
+import org.bukkit.util.StringUtil;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@Deprecated
+public class GCCommand extends VanillaCommand {
+
+    public GCCommand() {
+        super("tgc");
+        this.description = "Clear server memory.";
+        this.usageMessage = "/gc";
+        this.setPermission("torch.command.gc");
+    }
+
+    @Override
+    public boolean execute(CommandSender sender, String currentAlias, String[] args) {
+        if (!testPermission(sender)) return true;
+
+		long m1 = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+        System.gc();
+        System.runFinalization();
+        long m2 = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+        if (m2 < m1)
+			sender.sendMessage(
+			ChatColor.GREEN + "Cleaned " + ChatColor.GOLD + ((m1 - m2) / 1024L / 1024L) + ChatColor.GREEN + "MB memory");
+          else
+            sender.sendMessage(ChatColor.GREEN + "Memory not require clean now.");
+
+        return true;
+    }
+}
-- 
2.8.2.windows.1

