From 5c0086a0db0cf03f02167519015ff19a51a05806 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 19:37:04 +0800
Subject: [PATCH] Fix optimize DataPalleteBlock and getType


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 73fe093..65a3d57 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -806,7 +806,7 @@ public abstract class World implements IBlockAccess {
     // Paper start - reduces need to do isLoaded before getType
     public IBlockData getTypeIfLoaded(BlockPosition blockposition) {
         if (captureTreeGeneration) return findCaptured(blockposition); // use seperate findCaptured() method
-        Chunk chunk = ((ChunkProviderServer) this.chunkProvider).getChunkIfLoaded(x >> 4, z >> 4);
+        Chunk chunk = this.getChunkIfLoaded(blockposition);
         if (chunk != null) {
             return chunk.getBlockData(x, y, z);
         }
-- 
2.8.2.windows.1

