From 2b187217de5f77b1b7c5a68144fc1c55d1f1343d Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 19:45:30 +0800
Subject: [PATCH] Fix missed DataPalleteBlock


diff --git a/src/main/java/net/minecraft/server/DataPaletteBlock.java b/src/main/java/net/minecraft/server/DataPaletteBlock.java
index 109f877..4fad127 100644
--- a/src/main/java/net/minecraft/server/DataPaletteBlock.java
+++ b/src/main/java/net/minecraft/server/DataPaletteBlock.java
@@ -6,7 +6,7 @@ public class DataPaletteBlock implements DataPaletteExpandable {
 
     private static final DataPalette d = new DataPaletteGlobal();
     protected static final IBlockData a = Blocks.AIR.getBlockData();
-    // protected DataBits b; // Paper - nope
+    protected DataBits b;
 	private final short[] data = new short[4096]; // Paper
     protected DataPalette c;
     private int e = 0;
@@ -33,7 +33,7 @@ public class DataPaletteBlock implements DataPaletteExpandable {
             }
 
             this.c.a(DataPaletteBlock.a);
-            // this.b = new DataBits(this.e, 4096); // Paper
+            this.b = new DataBits(this.e, 4096); // Paper
         }
     }
 
@@ -43,10 +43,8 @@ public class DataPaletteBlock implements DataPaletteExpandable {
 
         this.b(i);
 
-        // Paper start - DataBits -> data
-        for (int j = 0; j < 4096; ++j) {
-            IBlockData iblockdata1 = datapalette.a(data[j]);
-            // Paper end
+        for (int j = 0; j < databits.b(); ++j) {
+            IBlockData iblockdata1 = datapalette.a(databits.a(j));
 
             if (iblockdata1 != null) {
                 this.setBlockIndex(j, iblockdata1);
@@ -80,21 +78,11 @@ public class DataPaletteBlock implements DataPaletteExpandable {
 
         return iblockdata == null ? DataPaletteBlock.a : iblockdata;
     }
-	
-	// Paper start - serialization method
-    public DataBits toMojangBits() {
-        DataBits bits = new DataBits(this.e, 4096);
-        for (int i = 0; i < 4096; i++) {
-            bits.a(i, this.c.a(Block.REGISTRY_ID.fromId(data[i])));
-        }
-        return bits;
-    }
-    // Paper end
 
     public void b(PacketDataSerializer packetdataserializer) {
         packetdataserializer.writeByte(this.e);
         this.c.b(packetdataserializer);
-        packetdataserializer.a(toMojangBits().a()); // Paper - serialize
+        packetdataserializer.a(this.b.a());
     }
 
     @Nullable
@@ -150,7 +138,7 @@ public class DataPaletteBlock implements DataPaletteExpandable {
     }
 
     public int a() {
-        return 1 + this.c.a() + PacketDataSerializer.a(this.toMojangBits().b()) + this.toMojangBits().a().length * 8;
+        return 1 + this.c.a() + PacketDataSerializer.a(this.b.b()) + this.b.a().length * 8;
         // Paper end
     }
 }
-- 
2.8.2.windows.1

