From 39272cd9a4f26441e4d803a327913d550f13c0a8 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 04:14:07 +0800
Subject: [PATCH] Optimize Idle Furnaces, miscs


diff --git a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
index 2c3353a..afc8c88 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
@@ -22,6 +22,7 @@ public abstract class MobSpawnerAbstract {
     private int requiredPlayerRange = 16;
     private int spawnRange = 4;
     private int tickDelay = 0; // Paper
+	private int spawningPenalty = 0; // Migot
 
     public MobSpawnerAbstract() {}
 
@@ -124,6 +125,15 @@ public abstract class MobSpawnerAbstract {
                 if (flag) {
                     this.i();
                 }
+				// Migot start
+                else {
+                    this.spawnDelay += this.spawningPenalty;
+                    if(this.spawningPenalty < 40) {
+                        this.spawningPenalty++;
+                    }
+                }
+                // Migot end
+
             }
 
         }
@@ -137,6 +147,8 @@ public abstract class MobSpawnerAbstract {
 
             this.spawnDelay = this.minSpawnDelay + this.a().random.nextInt(i);
         }
+		
+		this.spawningPenalty = 0; // Migot
 
         if (!this.mobs.isEmpty()) {
             // this.a((MobSpawnerData) WeightedRandom.a(this.a().random, this.mobs)); // Hose
diff --git a/src/main/java/net/minecraft/server/TileEntityFurnace.java b/src/main/java/net/minecraft/server/TileEntityFurnace.java
index b908110..55be645 100644
--- a/src/main/java/net/minecraft/server/TileEntityFurnace.java
+++ b/src/main/java/net/minecraft/server/TileEntityFurnace.java
@@ -187,7 +187,7 @@ public class TileEntityFurnace extends TileEntityContainer implements ITickable,
                 if (!this.isBurning() && this.cookTime > 0) {
                     this.cookTime = MathHelper.clamp(this.cookTime - 2, 0, this.cookTimeTotal);
                 }
-            } else {
+            } else if (this.items[1] != null && this.items[1].getItem() != Items.BUCKET) { // Migot
                 // CraftBukkit start - Handle multiple elapsed ticks
                 if (this.burnTime <= 0 && this.canBurn()) { // CraftBukkit - == to <=
                     CraftItemStack fuel = CraftItemStack.asCraftMirror(this.items[1]);
-- 
2.8.2.windows.1

