From 959c7849dc7f253655bdf88a3f175e2f797bb918 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 04:28:53 +0800
Subject: [PATCH] Fix Disappearing Leash Hitches


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 384abf5..d27d102 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -349,7 +349,7 @@ public abstract class Entity implements ICommandListener {
         this.lastZ = this.locZ;
         this.lastPitch = this.pitch;
         this.lastYaw = this.yaw;
-        if (!this.world.isClientSide && this.world instanceof WorldServer) {
+        if (!this.world.isClientSide && this.world instanceof WorldServer && !(this instanceof EntityProjectile) && !(this instanceof EntityArrow) && !(this instanceof EntityFireball)) { // Migot
             this.world.methodProfiler.a("portal");
             if (this.ak) {
                 MinecraftServer minecraftserver = this.world.getMinecraftServer();
diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
index c9b0149..6085275 100644
--- a/src/main/java/net/minecraft/server/EntityInsentient.java
+++ b/src/main/java/net/minecraft/server/EntityInsentient.java
@@ -359,6 +359,12 @@ public abstract class EntityInsentient extends EntityLiving {
 
             nbttagcompound.set("Leash", nbttagcompound3);
         }
+		
+		// Migot start
+        else if(this.bF != null) {
+            nbttagcompound.set("Leash", this.bF);
+        }
+        // Migot end
 
         nbttagcompound.setBoolean("LeftHanded", this.cT());
         if (this.bB != null) {
@@ -1129,11 +1135,13 @@ public abstract class EntityInsentient extends EntityLiving {
                     EntityLiving entityliving = (EntityLiving) iterator.next();
 
                     if (entityliving.getUniqueID().equals(uuid)) {
-                        this.leashHolder = entityliving;
+                        // this.leashHolder = entityliving;
+						this.setLeashHolder(entityliving, true); // Migot
                         break;
                     }
                 }
-            } else if (this.bF.hasKeyOfType("X", 99) && this.bF.hasKeyOfType("Y", 99) && this.bF.hasKeyOfType("Z", 99)) {
+            }
+            if (this.bF == null && this.bF.hasKeyOfType("X", 99) && this.bF.hasKeyOfType("Y", 99) && this.bF.hasKeyOfType("Z", 99)) { // Migot
                 BlockPosition blockposition = new BlockPosition(this.bF.getInt("X"), this.bF.getInt("Y"), this.bF.getInt("Z"));
                 EntityLeash entityleash = EntityLeash.b(this.world, blockposition);
 
@@ -1141,8 +1149,9 @@ public abstract class EntityInsentient extends EntityLiving {
                     entityleash = EntityLeash.a(this.world, blockposition);
                 }
 
-                this.leashHolder = entityleash;
-            } else {
+				this.setLeashHolder(entityleash, true); // Migot
+            }
+            if (this.leashHolder == null) { // Migot
                 this.world.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.UNKNOWN)); // CraftBukkit
                 this.unleash(false, true);
             }
-- 
2.8.2.windows.1

