From 646ad348b89d88c2e8f478d8e88168ee08de5d86 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 2 Jul 2016 22:13:43 +0800
Subject: [PATCH] Fix implement a new and improved in-memory UUID cache
 actually


diff --git a/src/main/java/com/destroystokyo/paper/profile/event/AsyncUUIDPreResolveEvent.java b/src/main/java/com/destroystokyo/paper/profile/event/AsyncUUIDPreResolveEvent.java
index 3dde2f4..ec1ab44 100644
--- a/src/main/java/com/destroystokyo/paper/profile/event/AsyncUUIDPreResolveEvent.java
+++ b/src/main/java/com/destroystokyo/paper/profile/event/AsyncUUIDPreResolveEvent.java
@@ -18,7 +18,7 @@ package com.destroystokyo.paper.profile.event;
      private final UUID id;
  
      public AsyncUUIDPreResolveEvent(UUID id) {
-         super(LookupCause.LOOKUP);
+         super(LookupCause.UUID);
          this.id = Preconditions.checkNotNull(id, "Null id");;
      }
  
-- 
2.8.2.windows.1

