From db25749859ccba0e174a41033514a2186d95d8c7 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 16:30:30 +0800
Subject: [PATCH] Fix var for entity


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index d164ece..6e819e1 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -3152,8 +3152,8 @@ public abstract class Entity implements ICommandListener {
 
         @Override
         protected void compute() {
-            am();
-            ak();
+			al();
+			aj();
             if (world.isClientSide) {
                 fireTicks = 0;
             } else if (fireTicks > 0) {
-- 
2.8.2.windows.1

