From 25b72b96debcbda81595b2295c963753d0462210 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 16:40:29 +0800
Subject: [PATCH] Try to fix threadsafe issue


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 6e819e1..ff739cf 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -3102,7 +3102,7 @@ public abstract class Entity implements ICommandListener {
         protected void compute() {
             collide_hose(entity);
         }
-
+	}
 	
 	class move extends RecursiveAction {
         double d0, d1, d2;
@@ -3148,7 +3148,9 @@ public abstract class Entity implements ICommandListener {
 	
 	class U_sub extends RecursiveAction {
         
-        public U_sub() {}
+        public U_sub() {
+			
+		}
 
         @Override
         protected void compute() {
@@ -3172,5 +3174,5 @@ public abstract class Entity implements ICommandListener {
             }
         }
     }
-	
-}
+
+}
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 281f48e..0428602 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -26,6 +26,7 @@ import org.spigotmc.AsyncCatcher;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Queues;
 import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.Queue;
 
 public class PlayerChunkMap {
@@ -55,8 +56,8 @@ public class PlayerChunkMap {
 	private final Map<Long, PlayerChunk> e = Maps.newConcurrentMap();
     // private final Set<PlayerChunk> f = AsyncCatcher.catchAsyncUsage(Sets.newHashSet(), "set access PlayerChunkMap.f");
 	private final Set<PlayerChunk> f = Sets.newConcurrentHashSet();
-    private final List<PlayerChunk> g = AsyncCatcher.catchAsyncUsage(Lists.newLinkedList(), "list access PlayerChunkMap.g");
-    private final List<PlayerChunk> h = AsyncCatcher.catchAsyncUsage(Lists.newLinkedList(), "list access PlayerChunkMap.h");
+    private final List<PlayerChunk> g = Lists.ConcurrentLinkedDeque();
+    private final List<PlayerChunk> h = Lists.ConcurrentLinkedDeque();
     // private final List<PlayerChunk> i = AsyncCatcher.catchAsyncUsage(Lists.newArrayList(), "list access PlayerChunkMap.i");
 	private final Queue<PlayerChunk> i = Queues.newConcurrentLinkedQueue();
     // TacoSpigot end
-- 
2.8.2.windows.1

