From bafffb1c9e06251eaffe349b2ea7d8476c0b8d3b Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 18:28:41 +0800
Subject: [PATCH] Check-for-lava-only-once-per-tick


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index b9200b9..f02ef88 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -76,6 +76,8 @@ public abstract class Entity implements ICommandListener {
     public boolean i;
     public final List<Entity> passengers;
     protected int j;
+	private boolean isInLava;
+    private int lastLavaCheck = Integer.MIN_VALUE;
     private Entity at;
     public boolean attachedToPlayer;
     public World world;
@@ -1451,7 +1453,12 @@ public abstract class Entity implements ICommandListener {
     }
 
     public boolean an() {
-        return this.world.a(this.getBoundingBox().grow(-0.10000000149011612D, -0.4000000059604645D, -0.10000000149011612D), Material.LAVA);
+		int currentTick = MinecraftServer.currentTick;
+        if (this.lastLavaCheck != currentTick) {
+            this.lastLavaCheck = currentTick;
+            this.isInLava = this.world.a(this.getBoundingBox().grow(-0.10000000149011612D, -0.4000000059604645D, -0.10000000149011612D), Material.LAVA);
+        }
+        return this.isInLava;
     }
 
     public void a(float f, float f1, float f2) {
diff --git a/src/main/java/net/minecraft/server/StructureGenerator.java b/src/main/java/net/minecraft/server/StructureGenerator.java
index 150fb46..aa4bea4 100644
--- a/src/main/java/net/minecraft/server/StructureGenerator.java
+++ b/src/main/java/net/minecraft/server/StructureGenerator.java
@@ -203,7 +203,7 @@ public abstract class StructureGenerator extends WorldGenBase {
     protected synchronized void a(World world) { // CraftBukkit - synchronized
         if (this.a == null) {
             // Spigot Start
-			if ( world.spigotConfig.saveStructureInfo && ( !this.a().equals( "Mineshaft" ) || world.spigotConfig.saveMineshaftStructureInfo ) ) // Migot
+			if ( world.spigotConfig.saveStructureInfo && ( !this.a().equals( "Mineshaft" ) || world.spigotConfig.saveMineshaftStructureInfo )){ // Migot
                 this.a = (PersistentStructure) world.a(PersistentStructure.class, this.a());
             } else {
                 this.a = new PersistentStructure(this.a());
-- 
2.8.2.windows.1

