From 4587025a3440e0a107abe6e7e3dc99470985227c Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 17:58:23 +0800
Subject: [PATCH] Reallow-Saving-of-Mineshaft-Info


diff --git a/src/main/java/net/minecraft/server/StructureGenerator.java b/src/main/java/net/minecraft/server/StructureGenerator.java
index 8b6294e..7df16e9 100644
--- a/src/main/java/net/minecraft/server/StructureGenerator.java
+++ b/src/main/java/net/minecraft/server/StructureGenerator.java
@@ -203,7 +203,7 @@ public abstract class StructureGenerator extends WorldGenBase {
     protected synchronized void a(World world) { // CraftBukkit - synchronized
         if (this.a == null) {
             // Spigot Start
-            if (world.spigotConfig.saveStructureInfo && !this.a().equals( "Mineshaft" )) {
+			if ( world.spigotConfig.saveStructureInfo && ( !this.a().equals( "Mineshaft" ) || world.spigotConfig.saveMineshaftStructureInfo ) ) // Migot
                 this.a = (PersistentStructure) world.a(PersistentStructure.class, this.a());
             } else {
                 this.a = new PersistentStructure(this.a());
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 756a1e2..397ac6a 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -304,10 +304,15 @@ public class SpigotWorldConfig
     }
 
     public boolean saveStructureInfo;
+	public boolean saveMineshaftStructureInfo; // Migot
     private void structureInfo()
     {
         saveStructureInfo = getBoolean( "save-structure-info", true );
         log( "Structure Info Saving: " + saveStructureInfo );
+		// Migot start
+        saveMineshaftStructureInfo = getBoolean( "save-mineshaft-structure-info", false);
+        log( "Mineshaft Structure Info Saving: " + saveMineshaftStructureInfo );
+        // Migot end
         if ( !saveStructureInfo )
         {
             log( "*** WARNING *** You have selected to NOT save structure info. This may cause structures such as fortresses to not spawn mobs!" );
-- 
2.8.2.windows.1

