From 047306c8428fa7035b732a27e4e6d046178e3fd9 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 17:31:39 +0800
Subject: [PATCH] SPIGOT-2481: Implement EnchantmentTarget.BREAKABLE


diff --git a/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java b/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
index a103e56..0b59cb0 100644
--- a/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
+++ b/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
@@ -47,6 +47,8 @@ public class CraftEnchantment extends Enchantment {
             return EnchantmentTarget.BOW;
         case FISHING_ROD:
             return EnchantmentTarget.FISHING_ROD;
+		case BREAKABLE:
+            return EnchantmentTarget.BREAKABLE;
         default:
             return null;
         }
diff --git a/src/test/java/org/bukkit/PerMaterialTest.java b/src/test/java/org/bukkit/PerMaterialTest.java
index b727d1d..3a6c63e 100644
--- a/src/test/java/org/bukkit/PerMaterialTest.java
+++ b/src/test/java/org/bukkit/PerMaterialTest.java
@@ -21,6 +21,7 @@ import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 import org.junit.runners.Parameterized.Parameter;
 import org.junit.runners.Parameterized.Parameters;
+import org.bukkit.enchantments.EnchantmentTarget;
 
 import com.google.common.collect.Lists;
 import java.util.Map;
@@ -146,4 +147,13 @@ public class PerMaterialTest extends AbstractTestingBase {
             assertFalse(material.hasGravity());
         }
     }
+	
+	@Test
+    public void usesDurability() {
+        if (!material.isBlock()) {
+            assertThat(EnchantmentTarget.BREAKABLE.includes(material), is(CraftMagicNumbers.getItem(material).usesDurability()));
+        } else {
+            assertFalse(EnchantmentTarget.BREAKABLE.includes(material));
+        }
+    }
 }
-- 
2.8.2.windows.1

