From 6933985b134b40c4693b448941c883509c6e0cd3 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 03:44:17 +0800
Subject: [PATCH] MC-66089 and MC-100260 Fix explosion distance


diff --git a/src/main/java/net/minecraft/server/Explosion.java b/src/main/java/net/minecraft/server/Explosion.java
index 604cdfe..ee5cd2c 100644
--- a/src/main/java/net/minecraft/server/Explosion.java
+++ b/src/main/java/net/minecraft/server/Explosion.java
@@ -147,7 +147,7 @@ public class Explosion {
                             continue;
                         }
                         // CraftBukkit end
-                        double d14 = 1.0D;
+                        double d14 = d13; // SportBukkit - fix calculations above being thrown away for non-living entities
 
                         if (entity instanceof EntityLiving) {
                             d14 = EnchantmentProtection.a((EntityLiving) entity, d13);
-- 
2.8.2.windows.1

