From d6463aa96ed5144bb458cd9378bf0034b361f218 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 06:21:03 +0800
Subject: [PATCH] Fix passenger desync


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index acdf0c9..22feac7 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -2959,6 +2959,7 @@ public abstract class Entity implements ICommandListener {
         return null;
     }
 
+	public List<Entity> getPassengers() { return this.bv(); } // SportBukkit - alias for following method
     public List<Entity> bv() {
         return (List) (this.passengers.isEmpty() ? Collections.emptyList() : Lists.newArrayList(this.passengers));
     }
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 2afdd77..8c59e15 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -439,6 +439,12 @@ public class EntityTrackerEntry {
                             entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityEffect(this.tracker.getId(), mobeffect));
                         }
                     }
+					
+					// SportBukkit start - inform new viewer of our passengers
+                    if(!this.tracker.getPassengers().isEmpty()) {
+                        entityplayer.playerConnection.sendPacket(new PacketPlayOutMount(this.tracker));
+                    }
+                    // SportBukkit end
 
                     this.tracker.b(entityplayer);
                     entityplayer.d(this.tracker);
-- 
2.8.2.windows.1

