From 7db2879435080d3e682b256f6ccaa2de117b98bf Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 18:08:11 +0800
Subject: [PATCH] Improvements to BookMeta API actually


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index 7f2a30a..9d96f9d 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -178,6 +178,10 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
     public boolean hasPages() {
         return !pages.isEmpty();
     }
+	
+	public boolean hasGeneration() {
+        return generation != null;
+    }
 
     public String getTitle() {
         return this.title;
@@ -296,6 +300,9 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         if (hasPages()) {
             hash = 61 * hash + 17 * this.pages.hashCode();
         }
+		if (hasGeneration()) {
+            hash = 61 * hash + 19 * this.generation.hashCode();
+        }
         return original != hash ? CraftMetaBook.class.hashCode() ^ hash : hash;
     }
 
@@ -309,7 +316,8 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
 
             return (hasTitle() ? that.hasTitle() && this.title.equals(that.title) : !that.hasTitle())
                     && (hasAuthor() ? that.hasAuthor() && this.author.equals(that.author) : !that.hasAuthor())
-                    && (hasPages() ? that.hasPages() && this.pages.equals(that.pages) : !that.hasPages());
+                    && (hasPages() ? that.hasPages() && this.pages.equals(that.pages) : !that.hasPages())
+                    && (hasGeneration() ? that.hasGeneration() && this.generation.equals(that.generation) : !that.hasGeneration());
         }
         return true;
     }
-- 
2.8.2.windows.1

