From 792760337a1411a96cf270dd7decda0c167396cf Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 06:24:47 +0800
Subject: [PATCH] Run data converters on passenger entities


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1c71146..b0b4f95 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -154,6 +154,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         this.b = this.i();
         // this.convertable = new WorldLoaderServer(file); // CraftBukkit - moved to DedicatedServer.init
         this.dataConverterManager = dataconvertermanager;
+		this.dataConverterManager.a(DataConverterTypes.ENTITY, new org.bukkit.craftbukkit.DataInspectorEntityPassengers()); // SportBukkit
         // CraftBukkit start
         this.options = options;
         EntityItem.uglyHack++; // Paper - MC-99914 - ensure EntityItem loads before EntityPotion
diff --git a/src/main/java/org/bukkit/craftbukkit/DataInspectorEntityPassengers.java b/src/main/java/org/bukkit/craftbukkit/DataInspectorEntityPassengers.java
new file mode 100644
index 0000000..caaf944
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/DataInspectorEntityPassengers.java
@@ -0,0 +1,23 @@
+package org.bukkit.craftbukkit;
+
+import net.minecraft.server.DataConverter;
+import net.minecraft.server.DataConverterTypes;
+import net.minecraft.server.DataInspector;
+import net.minecraft.server.NBTTagCompound;
+import net.minecraft.server.NBTTagList;
+
+/**
+ * Run data converters on passenger entities
+ */
+public class DataInspectorEntityPassengers implements DataInspector {
+    @Override
+    public NBTTagCompound a(DataConverter converter, NBTTagCompound nbt, int version) {
+        if(nbt.hasKeyOfType("Passengers", 9)) {
+            final NBTTagList passengers = nbt.getList("Passengers", 10);
+            for(int i = 0; i < passengers.size(); i++) {
+                passengers.a(i, converter.a(DataConverterTypes.ENTITY, passengers.get(i), version));
+            }
+        }
+        return nbt;
+    }
+}
\ No newline at end of file
-- 
2.8.2.windows.1

