From 8a3c6e4b3c0a18c75ec908c5a3a503285a7d1456 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 18:38:25 +0800
Subject: [PATCH] Fix missed method for bukkit command


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index aa40b1c..03b7d4f 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -31,7 +31,7 @@ public class SimpleCommandMap implements CommandMap {
         register("bukkit", new VersionCommand("version"));
         register("bukkit", new ReloadCommand("reload"));
         register("bukkit", new PluginsCommand("plugins"));
-        register("bukkit", new co.aikar.timings.TimingsCommand("timings")); // Spigot
+        register("bukkit", new TimingsCommand("timings"));
     }
 
     public void setFallbackCommands() {
@@ -222,7 +222,7 @@ public class SimpleCommandMap implements CommandMap {
         Map<String, String[]> values = server.getCommandAliases();
 
         for (String alias : values.keySet()) {
-            if (alias.contains(" ")) { // Torch
+            if (alias.contains(" ")) {
                 server.getLogger().warning("Could not register alias " + alias + " because it contains illegal characters");
                 continue;
             }
@@ -258,4 +258,4 @@ public class SimpleCommandMap implements CommandMap {
             }
         }
     }
-}
+}
\ No newline at end of file
-- 
2.8.2.windows.1

