From 53343aa09d3c3be60b87ef8e1e67cf5839db3249 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 06:42:22 +0800
Subject: [PATCH] More complex logic when looking for nearby players


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 3efc9b1..11d24d0 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -3085,7 +3085,7 @@ public abstract class World implements IBlockAccess {
         for (int i = 0; i < this.players.size(); ++i) {
             EntityHuman entityhuman1 = (EntityHuman) this.players.get(i);
             // CraftBukkit start - Fixed an NPE
-            if (entityhuman1 == null || entityhuman1.dead) {
+            if (entityhuman1 == null || !entityhuman1.isAlive()) { // CraftBukkit allow for more complex logic by using the "is alive" method
                 continue;
             }
             // CraftBukkit end
-- 
2.8.2.windows.1

