From e20f64ee4f07a5f5869aa274c26b5b7d9c750ea8 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 18:57:03 +0800
Subject: [PATCH] Fix entity villager for bukkit


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 4135e72..42d2aa4 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -20,7 +20,7 @@ public final class Versioning {
 
                 result = properties.getProperty("version");
             } catch (IOException ex) {
-                Logger.getLogger(Versioning.class.getName()).log(Level.SEVERE, "Could not get Bukkit version!", ex);
+                Logger.getLogger(Versioning.class.getName()).log(Level.SEVERE, "Could not get Torch version!", ex);
             }
         }
 
diff --git a/src/main/java/org/bukkit/entity/Villager.java b/src/main/java/org/bukkit/entity/Villager.java
index 2be0cb0..57e5272 100644
--- a/src/main/java/org/bukkit/entity/Villager.java
+++ b/src/main/java/org/bukkit/entity/Villager.java
@@ -119,12 +119,6 @@ public interface Villager extends Ageable, NPC, InventoryHolder {
         BLACKSMITH(3),
         BUTCHER(4),
 		HUSK(5);
-		
-		private final boolean zombie;
-
-        private Profession(boolean zombie) {
-            this.zombie = zombie;
-        }
 
         /**
          * Returns if this profession can only be used by zombies.
-- 
2.8.2.windows.1

