From 4fc2093b37b9d698bae01d052e424a02b3fd0e6e Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 11 Jul 2016 20:09:14 +0800
Subject: [PATCH] Fix for lang2


diff --git a/src/main/java/org/bukkit/command/defaults/EnchantCommand.java b/src/main/java/org/bukkit/command/defaults/EnchantCommand.java
index 6bbef26..32d9822 100644
--- a/src/main/java/org/bukkit/command/defaults/EnchantCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/EnchantCommand.java
@@ -6,7 +6,7 @@ import java.util.List;
 import java.util.Map;
 
 import org.apache.commons.lang3.Validate;
-import org.apache.commons.lang3.WordUtils;
+import org.apache.commons.lang.WordUtils;
 import com.google.common.collect.ImmutableList;
 
 import org.bukkit.Bukkit;
-- 
2.8.2.windows.1

