From 53eb884698e9fe8d40790518af2063c5720701fc Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Wed, 20 Jul 2016 18:19:49 +0800
Subject: [PATCH] Less strict skull validation, also fix TNTPrimed head hight
 calculation


diff --git a/src/main/java/net/minecraft/server/EntityTNTPrimed.java b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
index 028d68d..7c04fa5 100644
--- a/src/main/java/net/minecraft/server/EntityTNTPrimed.java
+++ b/src/main/java/net/minecraft/server/EntityTNTPrimed.java
@@ -157,7 +157,7 @@ public class EntityTNTPrimed extends Entity {
     }
 
     public float getHeadHeight() {
-        return world.paperConfig.oldCannonBehaviors ? this.length / 2 : this.length / 16f; // Paper - Old TNT cannon behaviors
+        return world.paperConfig.oldCannonBehaviors ? this.length / 16F : 0.0F; // Paper - Old TNT cannon behaviors
     }
 
     public void setFuseTicks(int i) {
diff --git a/src/main/java/net/minecraft/server/ItemSkull.java b/src/main/java/net/minecraft/server/ItemSkull.java
index 72c902a..0ad4cbf 100644
--- a/src/main/java/net/minecraft/server/ItemSkull.java
+++ b/src/main/java/net/minecraft/server/ItemSkull.java
@@ -135,11 +135,17 @@ public class ItemSkull extends Item {
             boolean valid = true;
 
             NBTTagList textures = nbttagcompound.getCompound("SkullOwner").getCompound("Properties").getList("textures", 10); // Safe due to method contracts
-            for (int i = 0; i < textures.size(); i++) {
-                if (textures.get(i) instanceof NBTTagCompound && !((NBTTagCompound) textures.get(i)).hasKeyOfType("Signature", 8)) {
+            // Paper start - Less strict validation
+            for (NBTBase texture : textures.list) {
+                if (texture instanceof NBTTagCompound && !((NBTTagCompound) texture).hasKeyOfType("Signature", 8)) {
+                    if (((NBTTagCompound) texture).getString("Value").trim().length() > 0) {
+                        continue;
+                    }
+
                     valid = false;
                 }
             }
+			// Paper end
 
             if (!valid) {
                 nbttagcompound.remove("SkullOwner");
-- 
2.8.2.windows.1

