From 3a4a7fdffb0b0f4bcc272aaacdef5714f8ad00d7 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 03:35:32 +0800
Subject: [PATCH] Fix chunkio2 timings issue


diff --git a/src/main/java/org/bukkit/craftbukkit/chunkio2/impl/CraftChunkRegionLoader.java b/src/main/java/org/bukkit/craftbukkit/chunkio2/impl/CraftChunkRegionLoader.java
index e5ca6f1..760ec34 100644
--- a/src/main/java/org/bukkit/craftbukkit/chunkio2/impl/CraftChunkRegionLoader.java
+++ b/src/main/java/org/bukkit/craftbukkit/chunkio2/impl/CraftChunkRegionLoader.java
@@ -42,9 +42,9 @@ public class CraftChunkRegionLoader implements IChunkLoader {
     @Override
     @CheckForNull
     public Chunk a(@Nonnull World world, int x, int z) throws IOException {
-        world.timings.syncChunkLoadDataTimer.startTiming(); // Spigot
+        // world.timings.syncChunkLoadDataTimer.startTiming(); // Spigot
         Object[] data = loadChunk(world, x, z);
-        world.timings.syncChunkLoadDataTimer.stopTiming(); // Spigot
+        // world.timings.syncChunkLoadDataTimer.stopTiming(); // Spigot
         if (data != null) {
             Chunk chunk = (Chunk) data[0];
             NBTTagCompound nbttagcompound = (NBTTagCompound) data[1];
-- 
2.8.2.windows.1

