From 1378e0f93bc9b2c7fed51bf214087b64c66f153f Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 22 Jul 2016 03:49:54 +0800
Subject: [PATCH] Dont use hopper push-based by default.


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 92fe1ad..db60c35 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -417,6 +417,6 @@ public class PaperWorldConfig {
 	
 	public boolean isHopperPushBased;
     private void isHopperPushBased() {
-        isHopperPushBased = getBoolean("hopper.push-based", true);
+        isHopperPushBased = getBoolean("hopper.push-based", false);
     }
 }
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
index a82623b..2fb2a45 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
@@ -65,7 +65,7 @@ public class TacoSpigotWorldConfig {
 
     public boolean isHopperPushBased;
     private void isHopperPushBased() {
-        isHopperPushBased = getBoolean("hopper.push-based", true); // Torch - default use
+        isHopperPushBased = getBoolean("hopper.push-based", false);
     }
 
     public boolean optimizeArmorStandMovement;
-- 
2.8.2.windows.1

