From 4414c5fb1ad58a67933ec116b690d0f157e83f32 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 03:51:04 +0800
Subject: [PATCH] Fix missed nbt storage


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 7228c42..5ecadc2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -41,7 +41,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
         this.server = server;
         this.profile = ProfileUtils.toMojang(profile);
         this.storage = server.console.worlds.get(0).getDataManager().getPlayerFileData(); // Paper
-        this.paperProfile = ProfileUtils.toPaper(profile);
+        this.paperProfile = profile;
     }
 
     public AccountProfile getAccount() {
-- 
2.8.2.windows.1

