From 98b52daf101b8d14f6af310a198133b10a062f5d Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 23 Jul 2016 23:00:29 +0800
Subject: [PATCH] Enhance Beacon Block API


diff --git a/src/main/java/net/minecraft/server/TileEntityBeacon.java b/src/main/java/net/minecraft/server/TileEntityBeacon.java
index 71e2fe2..bb740f1 100644
--- a/src/main/java/net/minecraft/server/TileEntityBeacon.java
+++ b/src/main/java/net/minecraft/server/TileEntityBeacon.java
@@ -42,9 +42,9 @@ public class TileEntityBeacon extends TileEntityContainer implements ITickable,
 	private MobEffectList primary;
 	private MobEffectList secondary;
     @Nullable
-    private MobEffectList l;
+    public MobEffectList l; // PAIL: private -> public
     @Nullable
-    private MobEffectList m;
+    public MobEffectList m; // PAIL: private -> public
     private ItemStack inventorySlot;
     private String o;
     // CraftBukkit start - add fields and methods
@@ -72,14 +72,11 @@ public class TileEntityBeacon extends TileEntityContainer implements ITickable,
     }
 	
 	public PotionEffect getPrimaryEffect() {
-        return CraftPotionUtil.toBukkit(new MobEffect(this.l, getLevel(), getAmplification(), true, true));
+       return (this.l != null) ? CraftPotionUtil.toBukkit(new MobEffect(this.l, getLevel(), getAmplification(), true, true)) : null;
     }
 
     public PotionEffect getSecondaryEffect() {
-        if (hasSecondaryEffect()) {
-            return CraftPotionUtil.toBukkit(new MobEffect(this.m, getLevel(), getAmplification(), true, true));
-        }
-        return null;
+        return (hasSecondaryEffect()) ? CraftPotionUtil.toBukkit(new MobEffect(this.m, getLevel(), getAmplification(), true, true)) : null;
     }
     // CraftBukkit end
 
diff --git a/src/main/java/org/bukkit/block/Beacon.java b/src/main/java/org/bukkit/block/Beacon.java
index f5311a5..32cb742 100644
--- a/src/main/java/org/bukkit/block/Beacon.java
+++ b/src/main/java/org/bukkit/block/Beacon.java
@@ -4,6 +4,7 @@ import java.util.Collection;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.inventory.InventoryHolder;
 import org.bukkit.potion.PotionEffect;
+import org.bukkit.potion.PotionEffectType;
 
 /**
  * Represents a beacon.
@@ -28,14 +29,30 @@ public interface Beacon extends BlockState, InventoryHolder {
     /**
      * Returns the primary effect set on the beacon
      *
-     * @return the primary effect
+     * @return the primary effect or null if not set
      */
     PotionEffect getPrimaryEffect();
 
     /**
+	 * Set the primary effect on this beacon, or null to clear.
+     *
+     * @param effect new primary effect
+     */
+    void setPrimaryEffect(PotionEffectType effect);
+
+    /**
      * Returns the secondary effect set on the beacon.
      *
      * @return the secondary effect or null if no secondary effect
      */
     PotionEffect getSecondaryEffect();
+	
+	/**
+     * Set the secondary effect on this beacon, or null to clear. Note that tier
+     * must be >= 4 for this effect to be active.
+     *
+     * @param effect desired secondary effect
+     */
+    void setSecondaryEffect(PotionEffectType effect);
+	
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBeacon.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBeacon.java
index 7a2a408..4ff4909 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBeacon.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBeacon.java
@@ -12,6 +12,9 @@ import org.bukkit.block.Beacon;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.inventory.CraftInventoryBeacon;
 import org.bukkit.inventory.Inventory;
+import net.minecraft.server.MobEffectList;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
+import org.bukkit.potion.PotionEffectType;
 
 public class CraftBeacon extends CraftBlockState implements Beacon {
     private final CraftWorld world;
@@ -71,10 +74,19 @@ public class CraftBeacon extends CraftBlockState implements Beacon {
     public PotionEffect getPrimaryEffect() {
         return beacon.getPrimaryEffect();
     }
+	
+	@Override
+    public void setPrimaryEffect(PotionEffectType effect) {
+        beacon.l = (effect != null) ? MobEffectList.fromId(effect.getId()) : null;
+    }
 
     @Override
     public PotionEffect getSecondaryEffect() {
         return beacon.getSecondaryEffect();
     }
 	
+	@Override
+    public void setSecondaryEffect(PotionEffectType effect) {
+        beacon.m = (effect != null) ? MobEffectList.fromId(effect.getId()) : null;
+    }
 }
\ No newline at end of file
-- 
2.8.2.windows.1

