From 3e724d260a56c6ec9c1fadf96c5f1fda1ea64be1 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 24 Jul 2016 18:05:13 +0800
Subject: [PATCH] Close inventory on player quit.


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 36fc983..f4503f8 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -402,6 +402,7 @@ public abstract class PlayerList {
 
         PlayerQuitEvent playerQuitEvent = new PlayerQuitEvent(cserver.getPlayer(entityplayer), "\u00A7e" + entityplayer.getName() + " left the game");
         cserver.getPluginManager().callEvent(playerQuitEvent);
+		entityplayer.closeInventory(); // Torch - close inventory on quit
         entityplayer.getBukkitEntity().disconnect(playerQuitEvent.getQuitMessage());
         // CraftBukkit end
 		
-- 
2.8.2.windows.1

