From 024b1500bc99e15e773a501c0f1bc7a22d3c5264 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 23 Jul 2016 00:41:27 +0800
Subject: [PATCH] SPIGOT-2537: Relax skull check even for nominally invalid
 skulls.


diff --git a/src/main/java/net/minecraft/server/ItemSkull.java b/src/main/java/net/minecraft/server/ItemSkull.java
index 0ad4cbf..1395152 100644
--- a/src/main/java/net/minecraft/server/ItemSkull.java
+++ b/src/main/java/net/minecraft/server/ItemSkull.java
@@ -132,24 +132,20 @@ public class ItemSkull extends Item {
             return true;
         } else {
 			// CraftBukkit start
-            boolean valid = true;
 
             NBTTagList textures = nbttagcompound.getCompound("SkullOwner").getCompound("Properties").getList("textures", 10); // Safe due to method contracts
             // Paper start - Less strict validation
             for (NBTBase texture : textures.list) {
-                if (texture instanceof NBTTagCompound && !((NBTTagCompound) texture).hasKeyOfType("Signature", 8)) {
-                    if (((NBTTagCompound) texture).getString("Value").trim().length() > 0) {
+				if (texture instanceof NBTTagCompound && !((NBTTagCompound) texture).hasKeyOfType("Signature", 8) && ((NBTTagCompound) texture).getString("Value").trim().isEmpty()) {
+                    nbttagcompound.remove("SkullOwner");
+                    break;
+					/* if (((NBTTagCompound) texture).getString("Value").trim().length() > 0) {
                         continue;
-                    }
-
-                    valid = false;
+                    } */
                 }
             }
 			// Paper end
 
-            if (!valid) {
-                nbttagcompound.remove("SkullOwner");
-            }
             // CraftBukkit end
             return false;
         }
-- 
2.8.2.windows.1

