From ef84388b38dc8a001e8f1962037580ec19577eb3 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 05:48:18 +0800
Subject: [PATCH] Fix nbt storage issue


diff --git a/src/main/java/net/minecraft/server/TileEntityFurnace.java b/src/main/java/net/minecraft/server/TileEntityFurnace.java
index 8dbfcc0..e12a8ed 100644
--- a/src/main/java/net/minecraft/server/TileEntityFurnace.java
+++ b/src/main/java/net/minecraft/server/TileEntityFurnace.java
@@ -252,7 +252,7 @@ public class TileEntityFurnace extends TileEntityContainer implements ITickable,
             return false;
         } else {
             ItemStack itemstack = RecipesFurnace.getInstance().getResult(this.items[0]);
-			if(!this.world.isLoaded(this.position) || ((WorldServer) this.world).chunkProviderServer.unloadQueue.contains(this.position.getX() >> 4, this.position.getZ() >> 4)) { return false; } // Migot
+			if(!this.world.isLoaded(this.position)) { return false; } // Migot
 
             // CraftBukkit - consider resultant count instead of current count
             return itemstack == null ? false : (this.items[2] == null ? true : (!this.items[2].doMaterialsMatch(itemstack) ? false : (this.items[2].count + itemstack.count <= this.getMaxStackSize() && this.items[2].count < this.items[2].getMaxStackSize() ? true : this.items[2].count + itemstack.count <= itemstack.getMaxStackSize())));
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 5ecadc2..813ae95 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -31,7 +31,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     protected CraftOfflinePlayer(CraftServer server, GameProfile profile) {
         this.server = server;
         this.profile = profile;
-        this.storage = server.console.worlds.get(0).getDataManager().getPlayerFileData(); // Paper
+        this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
 		// Paper start - store our profile
         this.paperProfile = ProfileUtils.toPaper(profile);
     }
@@ -40,7 +40,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     protected CraftOfflinePlayer(CraftServer server, AccountProfile profile) {
         this.server = server;
         this.profile = ProfileUtils.toMojang(profile);
-        this.storage = server.console.worlds.get(0).getDataManager().getPlayerFileData(); // Paper
+        this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
         this.paperProfile = profile;
     }
 
-- 
2.8.2.windows.1

