From 2072199d7164697513e07a40b6347d43e83a2bb1 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Mon, 25 Jul 2016 17:43:32 +0800
Subject: [PATCH] Expose BungeeCord Setting to API of Spigot


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 011cb99..bd4672e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1924,6 +1924,12 @@ public final class CraftServer implements Server {
         return entity.getBukkitEntity();
     }
 	
+	// PaperDragon start - Expose BungeeCord Setting
+    public boolean isBungeeEnabled() {
+        return org.spigotmc.SpigotConfig.bungee;
+    }
+    // PaperDragon end
+	
 	@Override
     public boolean isDebugLoggingEnabled() {
         return ((DedicatedServer) this.console).propertyManager.getBoolean("debug", false);
-- 
2.8.2.windows.1

