From 8c6907355488dc30d4cd12d50f0b9a879d50e352 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 05:48:18 +0800
Subject: [PATCH] Fix get uuid


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 2f0c87f..1631033 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -23,13 +23,12 @@ import org.bukkit.configuration.serialization.SerializableAs;
 import org.bukkit.entity.Player;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
-import net.minecraft.server.IPlayerFileData; // Paper
 
 @SerializableAs("Player")
 public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializable {
     private final GameProfile profile;
     private final CraftServer server;
-    private final IPlayerFileData storage; // Paper
+    private final WorldNBTStorage storage; // Paper
 
     protected CraftOfflinePlayer(CraftServer server, GameProfile profile) {
         this.server = server;
@@ -176,6 +175,10 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
 
         return this.getUniqueId().equals(other.getUniqueId());
     }
+	
+	private File getDataFile() {
+        return new File(storage.getPlayerDir(), getUniqueId() + ".dat");
+    }
 
     @Override
     public int hashCode() {
@@ -211,7 +214,8 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
             if (data.hasKey("firstPlayed")) {
                 return data.getLong("firstPlayed");
             } else {
-                return storage.getLastModified(getUniqueId()); // Paper
+                File file = getDataFile();
+                return file.lastModified();
             }
         } else {
             return 0;
@@ -228,7 +232,8 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
             if (data.hasKey("lastPlayed")) {
                 return data.getLong("lastPlayed");
             } else {
-                return storage.getLastModified(getUniqueId()); // Paper
+                File file = getDataFile();
+                return file.lastModified();
             }
         } else {
             return 0;
-- 
2.8.2.windows.1

