From 66c5d312b9cc71ffba4919675cc7f0fb8bd73df3 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 12 Jul 2016 14:08:04 +0800
Subject: [PATCH] Fix get uuid


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index 7d73dc1..538a370 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -16,8 +16,10 @@ import org.bukkit.Location;
 import org.bukkit.OfflinePlayer;
 import org.bukkit.Server;
 import net.minecraft.server.WorldNBTStorage;
+import net.minecraft.server.IPlayerFileData;
 import java.io.File;
 import net.minecraft.server.EntityPlayer;
+import net.minecraft.server.IDataManager;
 import org.bukkit.configuration.serialization.ConfigurationSerializable;
 import org.bukkit.configuration.serialization.SerializableAs;
 import org.bukkit.entity.Player;
@@ -28,12 +30,12 @@ import org.bukkit.plugin.Plugin;
 public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializable {
     private final GameProfile profile;
     private final CraftServer server;
-    private final WorldNBTStorage storage; // Paper
+    private final IDataManager storage; // Paper
 
     protected CraftOfflinePlayer(CraftServer server, GameProfile profile) {
         this.server = server;
         this.profile = profile;
-        this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
+        this.storage = server.console.worlds.get(0).getDataManager();
 		// Paper start - store our profile
         this.paperProfile = ProfileUtils.toPaper(profile);
     }
@@ -42,7 +44,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     protected CraftOfflinePlayer(CraftServer server, AccountProfile profile) {
         this.server = server;
         this.profile = ProfileUtils.toMojang(profile);
-        this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
+        this.storage = server.console.worlds.get(0).getDataManager();
         this.paperProfile = profile;
     }
 
@@ -175,10 +177,6 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
 
         return this.getUniqueId().equals(other.getUniqueId());
     }
-	
-	private File getDataFile() {
-        return new File(storage.getPlayerDir(), getUniqueId() + ".dat");
-    }
 
     @Override
     public int hashCode() {
@@ -214,8 +212,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
             if (data.hasKey("firstPlayed")) {
                 return data.getLong("firstPlayed");
             } else {
-                File file = getDataFile();
-                return file.lastModified();
+                return storage.getLastModified(getUniqueId()); // Paper
             }
         } else {
             return 0;
@@ -232,8 +229,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
             if (data.hasKey("lastPlayed")) {
                 return data.getLong("lastPlayed");
             } else {
-                File file = getDataFile();
-                return file.lastModified();
+                return storage.getLastModified(getUniqueId()); // Paper
             }
         } else {
             return 0;
-- 
2.8.2.windows.1

