From f3140930fa0e7606df6f67872290a88fbe871c6a Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 31 Jul 2016 15:04:11 +0800
Subject: [PATCH] Prevent a NPE for IO executor.


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index a29e0fa..675da1c 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -884,7 +884,9 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         MinecraftTimings.processQueueTimer.stopTiming(); // Spigot
 
         MinecraftTimings.chunkIOTickTimer.startTiming(); // Spigot
-        org.bukkit.craftbukkit.chunkio.ChunkIOExecutor.tick();
+		synchronized(this) {
+			org.bukkit.craftbukkit.chunkio.ChunkIOExecutor.tick();
+		}
         MinecraftTimings.chunkIOTickTimer.stopTiming(); // Spigot
 
         MinecraftTimings.timeUpdateTimer.startTiming(); // Spigot
-- 
2.8.2.windows.1

