From eaefee43d5770f1d41a409ac82cdedb172cd1308 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 5 Aug 2016 13:23:19 +0800
Subject: [PATCH] Fixes for message format


diff --git a/src/main/java/net/minecraft/server/MCUtil.java b/src/main/java/net/minecraft/server/MCUtil.java
index f1799ca..554e411 100644
--- a/src/main/java/net/minecraft/server/MCUtil.java
+++ b/src/main/java/net/minecraft/server/MCUtil.java
@@ -25,7 +25,7 @@ public final class MCUtil {
      * @return
      */
     public static IChatBaseComponent cmpFromMessage(String str) {
-        return IChatBaseComponent.ChatSerializer.a("{\"text\":\"" + REPLACE_QUOTES.matcher(str).replaceAll("\\\"") + "\"}");
+        return IChatBaseComponent.ChatSerializer.a("{\"text\":\"" + REPLACE_QUOTES.matcher(str).replaceAll("\\\\\"") + "\"}");
     }
 
     /**
diff --git a/src/main/java/net/minecraft/server/PacketDataSerializer.java b/src/main/java/net/minecraft/server/PacketDataSerializer.java
index 102f8d7..b2e9557 100644
--- a/src/main/java/net/minecraft/server/PacketDataSerializer.java
+++ b/src/main/java/net/minecraft/server/PacketDataSerializer.java
@@ -326,8 +326,8 @@ public class PacketDataSerializer extends ByteBuf {
     public PacketDataSerializer a(String s) {
         byte[] abyte = s.getBytes(Charsets.UTF_8);
 
-        if (abyte.length > 32767) {
-            throw new EncoderException("String too big (was " + s.length() + " bytes encoded, max " + 32767 + ")");
+        if (abyte.length > 40767) { // EMC
+            throw new EncoderException("String too big (was " + s.length() + " bytes encoded, max " + 40767 + ")"); // EMC
         } else {
             this.d(abyte.length);
             this.writeBytes(abyte);
-- 
2.8.2.windows.1

