From b0e6f81f07afd086045c5b5243275b216a4029b8 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 31 Jul 2016 02:11:45 +0800
Subject: [PATCH] Fix nbt load


diff --git a/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java b/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java
index b2d5254..e26c6f3 100644
--- a/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java
+++ b/src/main/java/net/minecraft/server/NBTCompressedStreamTools.java
@@ -1,7 +1,7 @@
 package net.minecraft.server;
 
-import java.io.BufferedInputStream;
-import java.io.BufferedOutputStream;
+import org.torch.util.io.FastBufferedInputStream;
+import org.torch.util.io.FastBufferedOutputStream;
 import java.io.DataInput;
 import java.io.DataInputStream;
 import java.io.DataOutput;
@@ -15,7 +15,7 @@ import java.util.zip.GZIPOutputStream;
 public class NBTCompressedStreamTools {
 
     public static NBTTagCompound a(InputStream inputstream) throws IOException {
-        DataInputStream datainputstream = new DataInputStream(new BufferedInputStream(new GZIPInputStream(inputstream)));
+        DataInputStream datainputstream = new DataInputStream(new FastBufferedInputStream(new GZIPInputStream(inputstream)));
 
         NBTTagCompound nbttagcompound;
 
@@ -29,7 +29,7 @@ public class NBTCompressedStreamTools {
     }
 
     public static void a(NBTTagCompound nbttagcompound, OutputStream outputstream) throws IOException {
-        DataOutputStream dataoutputstream = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(outputstream)));
+        DataOutputStream dataoutputstream = new DataOutputStream(new FastBufferedOutputStream(new GZIPOutputStream(outputstream)));
 
         try {
             a(nbttagcompound, (DataOutput) dataoutputstream);
diff --git a/src/main/java/net/minecraft/server/NBTTagCompound.java b/src/main/java/net/minecraft/server/NBTTagCompound.java
index 1b2cd19..11de8e8 100644
--- a/src/main/java/net/minecraft/server/NBTTagCompound.java
+++ b/src/main/java/net/minecraft/server/NBTTagCompound.java
@@ -10,10 +10,11 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.Map.Entry;
 import javax.annotation.Nullable;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class NBTTagCompound extends NBTBase {
 
-    public Map<String, NBTBase> map = Maps.newHashMap(); // Paper
+    public Map<String, NBTBase> map = new ConcurrentHashMap(); // Paper
 
     public NBTTagCompound() {}
 
-- 
2.8.2.windows.1

