From 27c66ed93ba6a45fbe9d350e07b77cb8121034ce Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 6 Aug 2016 18:07:46 +0800
Subject: [PATCH] SPIGOT-2574: BlockPopulators added twice.


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6a2c5d8..f293699 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -975,10 +975,6 @@ public final class CraftServer implements Server {
         internal.setSpawnFlags(true, true);
         console.worlds.add(internal);
 
-        if (generator != null) {
-            internal.getWorld().getPopulators().addAll(generator.getDefaultPopulators(internal.getWorld()));
-        }
-
         pluginManager.callEvent(new WorldInitEvent(internal.getWorld()));
         System.out.print("Preparing start region for level " + (console.worlds.size() - 1) + " (Seed: " + internal.getSeed() + ")");
 
-- 
2.8.2.windows.1

