From 8c4691207386d316fcb6de3582410af928b78b7c Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 7 Aug 2016 02:30:43 +0800
Subject: [PATCH] Stop help topics with no value printing nothing - Fixes
 BUKKIT-5623


diff --git a/src/main/java/org/bukkit/help/GenericCommandHelpTopic.java b/src/main/java/org/bukkit/help/GenericCommandHelpTopic.java
index 0f897de..2b7af07 100644
--- a/src/main/java/org/bukkit/help/GenericCommandHelpTopic.java
+++ b/src/main/java/org/bukkit/help/GenericCommandHelpTopic.java
@@ -42,14 +42,26 @@ public class GenericCommandHelpTopic extends HelpTopic {
         sb.append(ChatColor.GOLD);
         sb.append("Description: ");
         sb.append(ChatColor.WHITE);
-        sb.append(command.getDescription());
+		
+        if (command.getDescription().isEmpty()) {
+            sb.append(ChatColor.ITALIC);
+            sb.append("No information available.");
+        } else {
+            sb.append(command.getDescription());
+        }
 
         sb.append("\n");
 
         sb.append(ChatColor.GOLD);
         sb.append("Usage: ");
         sb.append(ChatColor.WHITE);
-        sb.append(command.getUsage().replace("<command>", name.substring(1)));
+        
+		if (command.getUsage().isEmpty()) {
+            sb.append(ChatColor.ITALIC);
+            sb.append("No information available.");
+        } else {
+            sb.append(command.getUsage().replace("<command>", name.substring(1)));
+        }
 
         if (command.getAliases().size() > 0) {
             sb.append("\n");
diff --git a/src/main/java/org/bukkit/help/IndexHelpTopic.java b/src/main/java/org/bukkit/help/IndexHelpTopic.java
new file mode 100644
index 0000000..e2fb729
--- /dev/null
+++ b/src/main/java/org/bukkit/help/IndexHelpTopic.java
@@ -0,0 +1,112 @@
+package org.bukkit.help;
+
+import org.bukkit.ChatColor;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.ConsoleCommandSender;
+import org.bukkit.entity.Player;
+import org.bukkit.util.ChatPaginator;
+
+import java.util.Collection;
+
+/**
+ * This help topic generates a list of other help topics. This class is useful
+ * for adding your own index help topics. To enforce a particular order, use a
+ * sorted collection.
+ * <p>
+ * If a preamble is provided to the constructor, that text will be displayed
+ * before the first item in the index.
+ */
+public class IndexHelpTopic extends HelpTopic {
+
+    protected String permission;
+    protected String preamble;
+    protected Collection<HelpTopic> allTopics;
+
+    public IndexHelpTopic(String name, String shortText, String permission, Collection<HelpTopic> topics) {
+        this(name, shortText, permission, topics, null);
+    }
+
+    public IndexHelpTopic(String name, String shortText, String permission, Collection<HelpTopic> topics, String preamble) {
+        this.name = name;
+        this.shortText = shortText;
+        this.permission = permission;
+        this.preamble = preamble;
+        setTopicsCollection(topics);
+    }
+
+    /**
+     * Sets the contents of the internal allTopics collection.
+     *
+     * @param topics The topics to set.
+     */
+    protected void setTopicsCollection(Collection<HelpTopic> topics) {
+        this.allTopics = topics;
+    }
+
+    public boolean canSee(CommandSender sender) {
+        if (sender instanceof ConsoleCommandSender) {
+            return true;
+        }
+        if (permission == null) {
+            return true;
+        }
+        return sender.hasPermission(permission);
+    }
+
+    @Override
+    public void amendCanSee(String amendedPermission) {
+        permission = amendedPermission;
+    }
+
+    public String getFullText(CommandSender sender) {
+        StringBuilder sb = new StringBuilder();
+
+        if (preamble != null) {
+            sb.append(buildPreamble(sender));
+            sb.append("\n");
+        }
+
+        for (HelpTopic topic : allTopics) {
+            if (topic.canSee(sender)) {
+                String lineStr = buildIndexLine(sender, topic).replace("\n", ". ");
+                if (sender instanceof Player && lineStr.length() > ChatPaginator.GUARANTEED_NO_WRAP_CHAT_PAGE_WIDTH) {
+                    sb.append(lineStr.substring(0, ChatPaginator.GUARANTEED_NO_WRAP_CHAT_PAGE_WIDTH - 3));
+                    sb.append("...");
+                } else {
+                    sb.append(lineStr);
+                }
+                sb.append("\n");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Builds the topic preamble. Override this method to change how the index
+     * preamble looks.
+     *
+     * @param sender The command sender requesting the preamble.
+     * @return The topic preamble.
+     */
+    protected String buildPreamble(CommandSender sender) {
+        return ChatColor.GRAY + preamble;
+    }
+
+    /**
+     * Builds individual lines in the index topic. Override this method to
+     * change how index lines are rendered.
+     *
+     * @param sender The command sender requesting the index line.
+     * @param topic  The topic to render into an index line.
+     * @return The rendered index line.
+     */
+    protected String buildIndexLine(CommandSender sender, HelpTopic topic) {
+        StringBuilder line = new StringBuilder();
+        line.append(ChatColor.GOLD);
+        line.append(topic.getName());
+        line.append(": ");
+        line.append(ChatColor.WHITE);
+        line.append(topic.getShortText().isEmpty() ? ChatColor.ITALIC + "No information available." : topic.getShortText());
+        return line.toString();
+    }
+}
-- 
2.8.2.windows.1

