From aec99f13c410522a606ed736681eb7d6318b248e Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 2 Aug 2016 18:43:41 +0800
Subject: [PATCH] [Bleeding] Fixes Index out of bound


diff --git a/src/main/java/net/minecraft/server/ChunkRegionLoader.java b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
index 07f83a9..5c092e5 100644
--- a/src/main/java/net/minecraft/server/ChunkRegionLoader.java
+++ b/src/main/java/net/minecraft/server/ChunkRegionLoader.java
@@ -185,7 +185,9 @@ public class ChunkRegionLoader implements IChunkLoader, IAsyncChunkSaver {
                 if (nbttagcompound != null) {
                     int attempts = 0; Exception laste = null; while (attempts++ < 5) { // Paper
                     try {
-                        this.b(chunkcoordintpair, nbttagcompound);
+						synchronized(this) {
+							this.b(chunkcoordintpair, nbttagcompound);
+						}
                         laste = null; break; // Paper
                     } catch (Exception exception) {
                         //ChunkRegionLoader.a.error("Failed to save chunk", exception); // Paper
-- 
2.8.2.windows.1

