From ddf6cf779c04c7241d0f591c0eff64004002c80d Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 04:00:07 +0800
Subject: [PATCH] Disable resource leak detector


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index bcc830f..1c71146 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -145,6 +145,10 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         this.V = minecraftsessionservice;
         this.W = gameprofilerepository;
         this.X = usercache;
+		
+		// SportBukkit - disable netty's resource leak detection
+        io.netty.util.ResourceLeakDetector.setEnabled(false);
+		
         // this.universe = file; // CraftBukkit
         // this.p = new ServerConnection(this); // Spigot
         this.b = this.i();
-- 
2.8.2.windows.1

