From 72a2dc464b38bbb8d980a2c7e21393c6dc24684c Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Tue, 9 Aug 2016 17:16:03 +0800
Subject: [PATCH] Demo screen API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index c395b42..2b6ddb1 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -16,6 +16,7 @@ import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.lang.Class;
 import java.util.Set;
 import java.util.UUID;
 import java.util.logging.Level;
@@ -1754,4 +1755,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public int getProtocolVersion() {
         return getHandle().protocolVersion;
     }
+	
+	// Torch start
+	public void trollDemoScreen() {
+        getHandle().playerConnection.sendPacket(new PacketPlayOutGameStateChange(5, 0));
+   }
+	// Torch end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f61d0a8..1a086e0 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1657,4 +1657,14 @@ public interface Player extends HumanEntity, Conversable, CommandSender, Offline
     Spigot spigot();
     // Spigot end
 	
+	// Torch start
+	/**
+     * Troll a Demo-mode screen for a player
+     * It's a minecraft hidden feature.
+     * color.
+     *
+     * @param player The name of player.
+     */
+    public void trollDemoScreen();
+	// Torch end
 }
-- 
2.8.2.windows.1

