From cc02984e114ce1cdfaf2cab9ee6e42ab88f3cd33 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 06:41:14 +0800
Subject: [PATCH] Fix invalid player death messages


diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 8aca102..9db9d68 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -414,7 +414,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
             }
         }
 
-        IChatBaseComponent chatmessage = this.getCombatTracker().getDeathMessage();
+        IChatBaseComponent chatmessage = damagesource == DamageSource.GENERIC ? damagesource.getLocalizedDeathMessage(this) : this.getCombatTracker().getDeathMessage(); // SportBukkit
 
         String deathmessage = chatmessage.toPlainText();
         org.bukkit.event.entity.PlayerDeathEvent event = CraftEventFactory.callPlayerDeathEvent(this, loot, deathmessage, keepInventory);
@@ -468,7 +468,8 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
         EntityLiving entityliving = this.bW();
 
-        if (entityliving != null) {
+        // CraftBukkit - can't have a combat tracked killer with a generic damage source
+        if (entityliving != null && damagesource != DamageSource.GENERIC) {
             EntityTypes.MonsterEggInfo entitytypes_monsteregginfo = (EntityTypes.MonsterEggInfo) EntityTypes.eggInfo.get(EntityTypes.b((Entity) entityliving));
 
             if (entitytypes_monsteregginfo != null) {
-- 
2.8.2.windows.1

