From a2480a949994e8e7045cf81a62cf79e32702b0b7 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Fri, 12 Aug 2016 00:47:37 +0800
Subject: [PATCH] Add bungee redirect compatible.


diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java
index 68b24a0..59f7e5c 100644
--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerHandshakeEvent.java
@@ -182,6 +182,7 @@ public class PlayerHandshakeEvent extends Event implements Cancellable {
 	 * @return {@code true} if authentication failed, {@code false} otherwise
 	 */
 	public boolean isFailed() {
+		if (de.minetick.MigotConfig.bungeeForwardingCompatible) { return false; }
 		return this.failed;
 	}
 
diff --git a/src/main/java/de/minetick/AutoSaveJob.java b/src/main/java/de/minetick/AutoSaveJob.java
index eb6b60a..680be0e 100644
--- a/src/main/java/de/minetick/AutoSaveJob.java
+++ b/src/main/java/de/minetick/AutoSaveJob.java
@@ -42,7 +42,7 @@ public class AutoSaveJob {
 			if (FileIOThread.a().isDone()) {
 				MinecraftTimings.worldSaveTimer.startTiming();
 				FileIOThread.a().setNoDelay(false);
-				RegionFileCache.a();
+				synchronized(this) { RegionFileCache.a(); }
 				Bukkit.getPluginManager().callEvent(new WorldSaveEvent(this.worldserver.getWorld()));
 				MinecraftTimings.worldSaveTimer.stopTiming();
 			} else {
diff --git a/src/main/java/de/minetick/MigotConfig.java b/src/main/java/de/minetick/MigotConfig.java
index a5fc72e..b5d87b0 100644
--- a/src/main/java/de/minetick/MigotConfig.java
+++ b/src/main/java/de/minetick/MigotConfig.java
@@ -46,6 +46,7 @@ public class MigotConfig {
 		enabledAutoSave = configuration.getBoolean("auto-save", false);
 		enabledForkPing = configuration.getBoolean("settings.fork-ping", false);
 		enabledFMLMonitor = configuration.getBoolean("settings.enable-fml-monitor", false);
+		bungeeForwardingCompatible = configuration.getBoolean("settings.bungee-redirect-compatible", false);
 	}
 
 	private FileConfiguration loadConfig(File file) throws IOException, InvalidConfigurationException {
@@ -68,8 +69,9 @@ public class MigotConfig {
 	}
 
 	// Settings - values
-	public static boolean enabledAutoSave = false;
-	public static boolean enabledForkPing = false;
-	public static boolean enabledFMLMonitor = false;
+	public static boolean enabledAutoSave;
+	public static boolean enabledForkPing;
+	public static boolean enabledFMLMonitor;
+	public static boolean bungeeForwardingCompatible;
 
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/HandshakeListener.java b/src/main/java/net/minecraft/server/HandshakeListener.java
index d169c74..6abee8f 100644
--- a/src/main/java/net/minecraft/server/HandshakeListener.java
+++ b/src/main/java/net/minecraft/server/HandshakeListener.java
@@ -3,6 +3,8 @@ package net.minecraft.server;
 // CraftBukkit start
 import java.net.InetAddress;
 
+import de.minetick.MigotConfig;
+
 public class HandshakeListener implements PacketHandshakingInListener {
 
 	private static final com.google.gson.Gson gson = new com.google.gson.Gson(); // Spigot
@@ -116,7 +118,7 @@ public class HandshakeListener implements PacketHandshakingInListener {
 				}
 				// Don't try and handle default logic if it's been handled by
 				// the event.
-				if (!handledByEvent && proxyLogicEnabled) {
+				if (!handledByEvent && proxyLogicEnabled && !MigotConfig.bungeeForwardingCompatible) {
 					// Paper end
 					// Spigot Start
 					// if (org.spigotmc.SpigotConfig.bungee) { // Paper -
@@ -138,6 +140,18 @@ public class HandshakeListener implements PacketHandshakingInListener {
 						b.spoofedProfile = gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
 					}
 				}
+				if (!handledByEvent && proxyLogicEnabled && MigotConfig.bungeeForwardingCompatible) {
+					String[] split = packethandshakinginsetprotocol.hostname.split("\00");
+					if (split.length == 3 || split.length == 4) {
+						packethandshakinginsetprotocol.hostname = split[0];
+						b.l = new java.net.InetSocketAddress(split[1],
+								((java.net.InetSocketAddress) b.getSocketAddress()).getPort());
+						b.spoofedUUID = com.mojang.util.UUIDTypeAdapter.fromString(split[2]);
+					}
+					if (split.length == 4) {
+						b.spoofedProfile = gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
+					}
+				}
 				// Spigot End
 				((LoginListener) this.b.i()).hostname = packethandshakinginsetprotocol.hostname + ":"
 						+ packethandshakinginsetprotocol.port; // CraftBukkit -
diff --git a/src/main/java/net/minecraft/server/RegionFile.java b/src/main/java/net/minecraft/server/RegionFile.java
index 2e26fdf..f751f41 100644
--- a/src/main/java/net/minecraft/server/RegionFile.java
+++ b/src/main/java/net/minecraft/server/RegionFile.java
@@ -87,9 +87,7 @@ public class RegionFile {
 			int j;
 
 			for (j = 0; j < i; ++j) {
-				synchronized (this) {
-					this.f.add(Boolean.valueOf(true));
-				}
+				this.f.add(Boolean.valueOf(true));
 			}
 
 			this.f.set(0, Boolean.valueOf(false));
@@ -294,9 +292,7 @@ public class RegionFile {
 
 					for (j2 = 0; j2 < k1; ++j2) {
 						this.c.write(RegionFile.a);
-						synchronized (this) {
-							this.f.add(Boolean.valueOf(false));
-						}
+						this.f.add(Boolean.valueOf(false));
 					}
 
 					this.g += 4096 * k1;
@@ -356,7 +352,7 @@ public class RegionFile {
 
 	public void c() throws IOException {
 		if (this.c != null) {
-			this.c.close();
+			synchronized(this) { this.c.close(); }
 		}
 
 	}
diff --git a/src/main/resources/configurations/migot.yml b/src/main/resources/configurations/migot.yml
index 2aa6d52..7476173 100644
--- a/src/main/resources/configurations/migot.yml
+++ b/src/main/resources/configurations/migot.yml
@@ -5,4 +5,6 @@
 
 auto-save: false
 settings:
-  fork-ping: false
\ No newline at end of file
+  fork-ping: false
+  enable-fml-monitor: false
+  bungee-redirect-compatible: false
\ No newline at end of file
-- 
2.8.2.windows.1

