From 9065122df732610b263e10fcac022a9c0b8cad49 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 20 Aug 2016 14:04:57 +0800
Subject: [PATCH] Enhanced ServerListPingEvent


diff --git a/src/main/java/net/minecraft/server/PacketStatusListener.java b/src/main/java/net/minecraft/server/PacketStatusListener.java
index 3fc0b0c..c8f43cd 100644
--- a/src/main/java/net/minecraft/server/PacketStatusListener.java
+++ b/src/main/java/net/minecraft/server/PacketStatusListener.java
@@ -67,7 +67,7 @@ public class PacketStatusListener implements PacketStatusInListener {
 
 				ServerListPingEvent() {
 					super(((InetSocketAddress) networkManager.getSocketAddress()).getAddress(),
-							minecraftServer.getMotd(), minecraftServer.getPlayerList().getMaxPlayers());
+minecraftServer.getMotd(), minecraftServer.getPlayerList().getMaxPlayers(), networkManager);
 				}
 
 				@Override
diff --git a/src/main/java/org/bukkit/event/server/ServerListPingEvent.java b/src/main/java/org/bukkit/event/server/ServerListPingEvent.java
index ca0873d..ea9a260 100644
--- a/src/main/java/org/bukkit/event/server/ServerListPingEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerListPingEvent.java
@@ -8,6 +8,8 @@ import org.bukkit.entity.Player;
 import org.bukkit.event.HandlerList;
 import org.bukkit.util.CachedServerIcon;
 
+import net.minecraft.server.NetworkManager;
+
 /**
  * Called when a server list ping is coming in. Displayed players can be checked
  * and removed by {@link #iterator() iterating} over this event.
@@ -19,6 +21,18 @@ public class ServerListPingEvent extends ServerEvent implements Iterable<Player>
 	private String motd;
 	private final int numPlayers;
 	private int maxPlayers;
+	private NetworkManager networkManager;
+	
+	// Torch start
+	public ServerListPingEvent(final InetAddress address, final String motd, final int maxPlayers, final NetworkManager networkManager) {
+		super();
+		this.numPlayers = MAGIC_PLAYER_COUNT;
+		this.address = address;
+		this.motd = motd;
+		this.maxPlayers = maxPlayers;
+		this.networkManager = networkManager;
+	}
+	// Torch end
 
 	public ServerListPingEvent(final InetAddress address, final String motd, final int numPlayers,
 			final int maxPlayers) {
@@ -58,6 +72,10 @@ public class ServerListPingEvent extends ServerEvent implements Iterable<Player>
 	public InetAddress getAddress() {
 		return address;
 	}
+	
+	public NetworkManager getNetworkManager() {
+		return networkManager;
+	}
 
 	/**
 	 * Get the message of the day message.
-- 
2.8.2.windows.1

