From e0d9d7200c72cedeecdec40779f6f33b10232507 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sat, 16 Jul 2016 06:54:03 +0800
Subject: [PATCH] Dont clear ItemStacks when amount is set to 0


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 6416609..113eb07 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -896,7 +896,7 @@ public class Chunk {
 
         for (int i = 0; i < j; ++i) {
             // CraftBukkit start
-            List<Entity> newList = Lists.newArrayList(this.entitySlices[i]);
+            Queue<Entity> newList = new ConcurrentLinkedQueue<Entity>(this.entitySlices[i]);
             java.util.Iterator<Entity> iter = newList.iterator();
             while (iter.hasNext()) {
                 Entity entity = iter.next();
@@ -1425,8 +1425,8 @@ public class Chunk {
     public Map<BlockPosition, TileEntity> getTileEntities() {
         return this.tileEntities;
     }
-
-    public Queue<Entity>[] getEntitySlices() { // Hose
+	
+	public Queue<Entity>[] getEntitySlices() { // Hose
         return this.entitySlices;
     }
 
diff --git a/src/main/java/net/minecraft/server/PacketDecoder.java b/src/main/java/net/minecraft/server/PacketDecoder.java
index fdf00d5..5687fae 100644
--- a/src/main/java/net/minecraft/server/PacketDecoder.java
+++ b/src/main/java/net/minecraft/server/PacketDecoder.java
@@ -24,21 +24,7 @@ public class PacketDecoder extends ByteToMessageDecoder {
     protected void decode(ChannelHandlerContext channelhandlercontext, ByteBuf bytebuf, List<Object> list) throws Exception {
         if (bytebuf.readableBytes() != 0) {
             PacketDataSerializer packetdataserializer = new PacketDataSerializer(bytebuf);
-			
-			// SportBukkit start - set protocol version
-            final NetworkManager networkManager = channelhandlercontext.pipeline().get(NetworkManager.class);
-            if(networkManager != null) {
-                packetdataserializer.protocolVersion = networkManager.protocolVersion;
-            }
-            // SportBukkit end
-			
             int i = packetdataserializer.g();
-			
-			// SportBukkit start - translate packet ID
-            i = ProtocolTranslator.packetId(true, packetdataserializer.protocolVersion, i);
-            if(i < 0) return;
-            // SportBukkit end
-			
             Packet packet = ((EnumProtocol) channelhandlercontext.channel().attr(NetworkManager.c).get()).a(this.c, i);
 
             if (packet == null) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java
index 393aa31..b81ffaf 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventoryPlayer.java
@@ -61,32 +61,6 @@ public class CraftInventoryPlayer extends CraftInventory implements org.bukkit.i
     public void setItemInHand(ItemStack stack) {
         setItemInMainHand(stack);
     }
-	
-	@Override
-    public ItemStack getItem(EquipmentSlot slot) {
-        switch(slot) {
-            case HAND: return getItemInMainHand();
-            case OFF_HAND: return getItemInOffHand();
-            case HEAD: return getHelmet();
-            case CHEST: return getChestplate();
-            case LEGS: return getLeggings();
-            case FEET: return getBoots();
-        }
-        throw new UnsupportedOperationException(slot.name());
-    }
-
-    @Override
-    public void setItem(EquipmentSlot slot, ItemStack stack) {
-        switch(slot) {
-            case HAND: setItemInMainHand(stack); return;
-            case OFF_HAND: setItemInOffHand(stack); return;
-            case HEAD: setHelmet(stack); return;
-            case CHEST: setChestplate(stack); return;
-            case LEGS: setLeggings(stack); return;
-            case FEET: setBoots(stack); return;
-        }
-        throw new UnsupportedOperationException(slot.name());
-    }
 
     @Override
     public void setItem(int index, ItemStack item) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index b397ccc..b67cec1 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -152,11 +152,7 @@ public final class CraftItemStack extends ItemStack {
         if (handle == null) {
             return;
         }
-        if (amount == 0) {
-            handle = null;
-        } else {
-            handle.count = amount;
-        }
+        handle.count = amount;
     }
 
     @Override
-- 
2.8.2.windows.1

