From 877e41bb65647bde97926edc1949b41a7ee20eef Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 28 Aug 2016 16:24:27 +0800
Subject: [PATCH] Make getEntity check more specific


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4352e0e..b260010 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1634,7 +1634,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
 			if (worldserver != null) {
 				Entity entity = worldserver.getEntity(uuid);
 
-				if (entity != null) {
+				if (entity != null && !entity.dead) {
 					return entity;
 				}
 			}
-- 
2.8.2.windows.1

