From d4966323c1fcc29186fe5998d268567a30300ee0 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Wed, 20 Jul 2016 18:16:43 +0800
Subject: [PATCH] Dont lookup game profiles that have no UUID and no name


diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index 9265148..92fe1ad 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -279,7 +279,7 @@ public class PaperWorldConfig {
 
     public boolean useHopperCheck;
     private void useHopperCheck() {
-        useHopperCheck = getBoolean("use-hopper-check", false);
+        useHopperCheck = getBoolean("use-hopper-check", false); // Torch - we still kepping this optional feature
     }
 
     public boolean allowLeashingUndeadHorse = false;
diff --git a/src/main/java/net/minecraft/server/UserCache.java b/src/main/java/net/minecraft/server/UserCache.java
index 157579a..7631f4c 100644
--- a/src/main/java/net/minecraft/server/UserCache.java
+++ b/src/main/java/net/minecraft/server/UserCache.java
@@ -86,7 +86,7 @@ public class UserCache {
         };
 
         gameprofilerepository.findProfilesByNames(new String[] { s}, Agent.MINECRAFT, profilelookupcallback);
-        if (!d() && agameprofile[0] == null) {
+        if (!d() && agameprofile[0] == null && !org.apache.commons.lang3.StringUtils.isBlank(s)) { // Paper - Don't lookup a profile with a blank name
             UUID uuid = EntityHuman.a(new GameProfile((UUID) null, s));
             GameProfile gameprofile = new GameProfile(uuid, s);
 
diff --git a/src/main/java/org/bukkit/command/defaults/GCCommand.java b/src/main/java/org/bukkit/command/defaults/GCCommand.java
index 51dad66..b5ac6d9 100644
--- a/src/main/java/org/bukkit/command/defaults/GCCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/GCCommand.java
@@ -19,7 +19,7 @@ public class GCCommand extends VanillaCommand {
 
     public GCCommand() {
         super("tgc");
-        this.description = "Clear server memory.";
+        this.description = "Cleans the server memory.";
         this.usageMessage = "/tgc";
         this.setPermission("bukkit.command.gc");
     }
@@ -34,9 +34,9 @@ public class GCCommand extends VanillaCommand {
         long m2 = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
         if (m2 < m1)
 			sender.sendMessage(
-			ChatColor.GOLD + "Cleaned " + ChatColor.GREEN + ((m1 - m2) / 1024L / 1024L) + "MB" + ChatColor.GOLD + " memory.");
+			ChatColor.GOLD + "GC - Cleaned " + ChatColor.GREEN + ((m1 - m2) / 1024L / 1024L) + "MB" + ChatColor.GOLD + " memory.");
           else
-            sender.sendMessage(ChatColor.GREEN + "Memory not require clean now.");
+            sender.sendMessage(ChatColor.GOLD + "GC - Memory not require clean now.");
 
         return true;
     }
-- 
2.8.2.windows.1

