From fb0265ee8ca29da175b6a6d5f1f93f160c4a5554 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Thu, 21 Jul 2016 15:59:31 +0800
Subject: [PATCH] SPIGOT-2414: Dontt disconnect if progressing


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 978e882..81fcee5 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -231,6 +231,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
     }
 
     public void disconnect(String s) {
+		if (this.processedDisconnect) return; // SPIGOT-2414
         // CraftBukkit start - fire PlayerKickEvent
         String leaveMessage = EnumChatFormat.YELLOW + this.player.getName() + " left the game.";
 
-- 
2.8.2.windows.1

