From dcf14ca3f08f937e6f15ecfdfebbe3e2361a2e99 Mon Sep 17 00:00:00 2001
From: SotrForgotten <i@omc.hk>
Date: Sun, 24 Jul 2016 13:56:37 +0800
Subject: [PATCH] Fix thread safe and sync issues


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 85f960a..1623d40 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -134,7 +134,9 @@ public class ChunkProviderServer implements IChunkProvider {
                 chunk = ChunkIOExecutor.syncChunkLoad(world, loader, this, i, j);
             }
 			*/ // Paper
-            chunk = ChunkIOExecutor.syncChunkLoad(world, chunkLoader, this, i, j); // Paper - skip chunkExists, expensive and performed by loading anyway.
+			synchronized(this) {
+				chunk = ChunkIOExecutor.syncChunkLoad(world, chunkLoader, this, i, j); // Paper - skip chunkExists, expensive and performed by loading anyway.
+			}
         }
 
         return chunk;
@@ -187,7 +189,9 @@ public class ChunkProviderServer implements IChunkProvider {
                 ChunkIOExecutor.queueChunkLoad(world, loader, this, i, j, runnable);
                 return null;
             } else {
-                chunk = ChunkIOExecutor.syncChunkLoad(world, loader, this, i, j);
+				synchronized(this) {
+					chunk = ChunkIOExecutor.syncChunkLoad(world, loader, this, i, j);
+				}
             }
         } else if (chunk == null && generate) {
             chunk = originalGetChunkAt(i, j);
diff --git a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
index 80eb01f..4f711f3 100644
--- a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
+++ b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
@@ -15,9 +15,7 @@ public class ChunkIOExecutor {
     private static final AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException> instance = new AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException>(new ChunkIOProvider(), BASE_THREADS);
 
     public static Chunk syncChunkLoad(World world, IChunkLoader loader, ChunkProviderServer provider, int x, int z) {
-		synchronized(instance) {
 			return instance.getSkipQueue(new QueuedChunk(x, z, loader, world, provider));
-		}
     }
 
     public static void queueChunkLoad(World world, IChunkLoader loader, ChunkProviderServer provider, int x, int z, Runnable runnable) {
diff --git a/src/main/java/org/bukkit/craftbukkit/util/AsynchronousExecutor.java b/src/main/java/org/bukkit/craftbukkit/util/AsynchronousExecutor.java
index 0bd34c3..193c362 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/AsynchronousExecutor.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/AsynchronousExecutor.java
@@ -118,7 +118,7 @@ public final class AsynchronousExecutor<P, T, C, E extends Throwable> {
             }
         }
 
-        synchronized void initSync() {
+        void initSync() {
             if (set(this, PENDING, STAGE_1_COMPLETE)) {
                 // If we succeed that variable switch, good as done
                 init();
-- 
2.8.2.windows.1

